<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>PuzzleMaker · Pseudoseq.jl</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../../assets/documenter.js"></script><script src="../../../siteinfo.js"></script><script src="../../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../../"><img src="../../../assets/logo.png" alt="Pseudoseq.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit">Pseudoseq.jl</span></div><form class="docs-search" action="../../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../../">Home</a></li><li><span class="tocitem">Manual</span><ul><li class="is-active"><a class="tocitem" href>PuzzleMaker</a><ul class="internal"><li><a class="tocitem" href="#The-MotifStitcher-1"><span>The MotifStitcher</span></a></li></ul></li><li><input class="collapse-toggle" id="menuitem-2-2" type="checkbox"/><label class="tocitem" for="menuitem-2-2"><span class="docs-label">Sequencing</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../sequencing/concepts/">Core concepts &amp; workflow</a></li><li><input class="collapse-toggle" id="menuitem-2-2-2" type="checkbox"/><label class="tocitem" for="menuitem-2-2-2"><span class="docs-label">Examples</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../sequencing/examples/pe-example/">Paired end reads</a></li><li><a class="tocitem" href="../../sequencing/examples/se-example/">Long single end reads</a></li><li><a class="tocitem" href="../../sequencing/examples/tg-example/">Tagged paired end reads</a></li></ul></li></ul></li></ul></li><li><span class="tocitem">API</span><ul><li><input class="collapse-toggle" id="menuitem-3-1" type="checkbox"/><label class="tocitem" for="menuitem-3-1"><span class="docs-label">Sequencing</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../../api/sequence/">sequence</a></li><li><a class="tocitem" href="../../../api/pool/">Molecule Pool</a></li><li><a class="tocitem" href="../../../api/reads/">Reads</a></li></ul></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Manual</a></li><li class="is-active"><a href>PuzzleMaker</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>PuzzleMaker</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/master/docs/src/man/puzzlemaker/concepts.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="PuzzleMaker:-*Core-concepts-and-basic-workflow*-1"><a class="docs-heading-anchor" href="#PuzzleMaker:-*Core-concepts-and-basic-workflow*-1">PuzzleMaker: <em>Core concepts &amp; basic workflow</em></a><a class="docs-heading-anchor-permalink" href="#PuzzleMaker:-*Core-concepts-and-basic-workflow*-1" title="Permalink"></a></h1><p><code>Pseudoseq</code> allows you to create puzzles; planned genomes and chromosome that have a certain set of features and peculiarities of interest.</p><p>The purpose for creating such puzzle genomes is not to recreate biology perfectly. The purpose is to create problems you understand fully (where the repeated content is, which positions are heterozygous and so on).</p><p>Using such genomes can help you both understand and develop an intuition of what current genome assembly tools are doing, and also to help design assembly tools, and perhaps even plan sequencing experiments and form hypotheses.</p><p><code>PuzzleMaker</code> is the <code>Pseudoseq</code> submodule that contains the functionality for doing this. This manual includes several examples showing how to create genomes with certain characteristics. But the core workflow, and important concepts are explained below.</p><h2 id="The-MotifStitcher-1"><a class="docs-heading-anchor" href="#The-MotifStitcher-1">The MotifStitcher</a><a class="docs-heading-anchor-permalink" href="#The-MotifStitcher-1" title="Permalink"></a></h2><p>The <code>MotifStitcher</code> is the type that is central to <code>PuzzleMaker</code>. It is by creating and interacting with a <code>MotifStitcher</code>, you create the plan for your puzzle in 3 simple steps. First you define motifs, then you decide on the order of motifs, and then you generate one or more instances of the puzzle by calling <code>make_puzzle</code> on your <code>MotifStitcher</code>.</p><p>Let&#39;s take a look at step 1.</p><h3 id=".-Define-Motifs-1"><a class="docs-heading-anchor" href="#.-Define-Motifs-1">1. Define Motifs</a><a class="docs-heading-anchor-permalink" href="#.-Define-Motifs-1" title="Permalink"></a></h3><p>The <code>MotifSticher</code> allows you to programmatically create puzzle haplotype sequences by first defining a set of motifs, motifs are shorter chunks of user specified or randomly generated sequence, that will be &quot;stitched&quot; together to form the final haplotype sequences.</p><p>There are three kinds of motif that can be added to the MotifStitcher.</p><ol><li>A random motif.</li><li>A fixed motif.</li><li>A sibling motif.</li></ol><h4 id="Random-motifs-1"><a class="docs-heading-anchor" href="#Random-motifs-1">Random motifs</a><a class="docs-heading-anchor-permalink" href="#Random-motifs-1" title="Permalink"></a></h4><p>A random motif is a motif that will change between calls to <code>make_puzzle</code>. The reason is so as you can trivially make multiple replicate puzzles with the same properties, but with different specific sequences.</p><p>You can add a random motif to a <code>MotifStitcher</code> by using the <code>add_motif!</code> method, and specifying only a length (in bp) for the random motif, for example:</p><pre><code class="language-julia-repl">julia&gt; ms = MotifStitcher()
Pseudoseq.PuzzleMaker.MotifStitcher(0, Dict{Int64,Pseudoseq.PuzzleMaker.RandomMotif}(), Dict{Int64,BioSequences.LongSequence{BioSequences.DNAAlphabet{2}}}(), Dict{Int64,Pseudoseq.PuzzleMaker.SiblingMotif}(), Array{Int64,1}[])

julia&gt; add_motif!(ms, 10_000)
Pseudoseq.PuzzleMaker.MotifStitcher(1, Dict(1 =&gt; Pseudoseq.PuzzleMaker.RandomMotif(10000, BioSequences.SamplerWeighted{BioSymbols.DNA}(BioSymbols.DNA[DNA_A, DNA_T, DNA_C, DNA_G], [0.25, 0.25, 0.25, 0.25]))), Dict{Int64,BioSequences.LongSequence{BioSequences.DNAAlphabet{2}}}(), Dict{Int64,Pseudoseq.PuzzleMaker.SiblingMotif}(), Array{Int64,1}[])</code></pre><p>By default, a random motif is constructed using a sampler that gives equal weighting to the four nucleotides. If you wanted more control over some of the nucleotide biases. You can construct a <code>RandomMotif</code> yourself, and provide it with your own nucleotide sampler:</p><pre><code class="language-julia-repl">julia&gt; ms = MotifStitcher()
Pseudoseq.PuzzleMaker.MotifStitcher(0, Dict{Int64,Pseudoseq.PuzzleMaker.RandomMotif}(), Dict{Int64,BioSequences.LongSequence{BioSequences.DNAAlphabet{2}}}(), Dict{Int64,Pseudoseq.PuzzleMaker.SiblingMotif}(), Array{Int64,1}[])

julia&gt; smp = SamplerWeighted(dna&quot;ACGT&quot;, [0.2, 0.3, 0.3,]) # Sampler biased toward GC.
BioSequences.SamplerWeighted{BioSymbols.DNA}(BioSymbols.DNA[DNA_A, DNA_C, DNA_G, DNA_T], [0.2, 0.3, 0.3, 0.19999999999999996])

julia&gt; add_motif!(ms, RandomMotif(10_000, smp)) # RandomMotif of 10_000 bp in length, using custom sampler.
Pseudoseq.PuzzleMaker.MotifStitcher(1, Dict(1 =&gt; Pseudoseq.PuzzleMaker.RandomMotif(10000, BioSequences.SamplerWeighted{BioSymbols.DNA}(BioSymbols.DNA[DNA_A, DNA_C, DNA_G, DNA_T], [0.2, 0.3, 0.3, 0.19999999999999996]))), Dict{Int64,BioSequences.LongSequence{BioSequences.DNAAlphabet{2}}}(), Dict{Int64,Pseudoseq.PuzzleMaker.SiblingMotif}(), Array{Int64,1}[])</code></pre><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>Only <code>SamplerWeighted{DNA}</code> types are accepted.</p></div></div><h4 id="Fixed-motifs-1"><a class="docs-heading-anchor" href="#Fixed-motifs-1">Fixed motifs</a><a class="docs-heading-anchor-permalink" href="#Fixed-motifs-1" title="Permalink"></a></h4><p>Unlike a random motif, a fixed motif has its sequence defined and constant over multiple calls of <code>make_puzzle</code>.</p><p>This is useful for situations where your puzzles must always include a certain DNA sequence. Perhaps one of biological interest or known to confuse a heuristic.</p><p>You add a fixed motif to a <code>MotifStitcher</code> simply by passing it a DNA sequence:</p><pre><code class="language-julia-repl">julia&gt; ms = MotifStitcher()
Pseudoseq.PuzzleMaker.MotifStitcher(0, Dict{Int64,Pseudoseq.PuzzleMaker.RandomMotif}(), Dict{Int64,BioSequences.LongSequence{BioSequences.DNAAlphabet{2}}}(), Dict{Int64,Pseudoseq.PuzzleMaker.SiblingMotif}(), Array{Int64,1}[])

julia&gt; add_motif!(ms, dna&quot;ATCGATCG&quot;)
Pseudoseq.PuzzleMaker.MotifStitcher(1, Dict{Int64,Pseudoseq.PuzzleMaker.RandomMotif}(), Dict(1 =&gt; ATCGATCG), Dict{Int64,Pseudoseq.PuzzleMaker.SiblingMotif}(), Array{Int64,1}[])</code></pre><h4 id="Sibling-motifs-1"><a class="docs-heading-anchor" href="#Sibling-motifs-1">Sibling motifs</a><a class="docs-heading-anchor-permalink" href="#Sibling-motifs-1" title="Permalink"></a></h4><p>A sibling motif is a motif that is randomly generated for each call of <code>make_puzzle</code> just like random motifs. However, unlike random motifs, a sibling motif is defined in terms of another motif already defined in the <code>MotifStitcher</code>.</p><p>To define a sibling motif, you specify an already existing motif. That motif&#39;s sequence forms the base sequence of the new sibling motif. To define a sibling motif you also need to provide a value that specifies the proportion of bases in the new sibling motif&#39;s sequence, that should differ in their nucleic acid from the base motifs sequence.</p><p>So sibling motifs then make it simple to define motifs that have a certain level of sequence similarity / homology / shared ancestry, with another motif. Creating portions of a simulated diploid genome might be one practical application of sibling motifs.</p><p>You add a sibling motif to the <code>MotifStitcher</code> by providing the <code>add_motif!</code> method with an <code>Pair{Int,Float64}</code>. where the integer is the ID of the chosen base motif already defined in the <code>MotifStitcher</code>, and the floating point number specifies the proportion of differing bases in the new motif&#39;s sequence:</p><pre><code class="language-julia-repl">julia&gt; ms = MotifStitcher()
Pseudoseq.PuzzleMaker.MotifStitcher(0, Dict{Int64,Pseudoseq.PuzzleMaker.RandomMotif}(), Dict{Int64,BioSequences.LongSequence{BioSequences.DNAAlphabet{2}}}(), Dict{Int64,Pseudoseq.PuzzleMaker.SiblingMotif}(), Array{Int64,1}[])

julia&gt; add_motif!(ms, 10_000) # A random first 10,000bp motif. Has ID = 1.
Pseudoseq.PuzzleMaker.MotifStitcher(1, Dict(1 =&gt; Pseudoseq.PuzzleMaker.RandomMotif(10000, BioSequences.SamplerWeighted{BioSymbols.DNA}(BioSymbols.DNA[DNA_A, DNA_T, DNA_C, DNA_G], [0.25, 0.25, 0.25, 0.25]))), Dict{Int64,BioSequences.LongSequence{BioSequences.DNAAlphabet{2}}}(), Dict{Int64,Pseudoseq.PuzzleMaker.SiblingMotif}(), Array{Int64,1}[])

julia&gt; add_motif!(ms, 1 =&gt; 0.01) # Add a sibling motif that will have ~10 bases which differ from motif #1. Will have ID = 2.
Pseudoseq.PuzzleMaker.MotifStitcher(2, Dict(1 =&gt; Pseudoseq.PuzzleMaker.RandomMotif(10000, BioSequences.SamplerWeighted{BioSymbols.DNA}(BioSymbols.DNA[DNA_A, DNA_T, DNA_C, DNA_G], [0.25, 0.25, 0.25, 0.25]))), Dict{Int64,BioSequences.LongSequence{BioSequences.DNAAlphabet{2}}}(), Dict(2 =&gt; Pseudoseq.PuzzleMaker.SiblingMotif(1, 0.01)), Array{Int64,1}[])</code></pre><h3 id=".-Specify-haplotypes-1"><a class="docs-heading-anchor" href="#.-Specify-haplotypes-1">2. Specify haplotypes</a><a class="docs-heading-anchor-permalink" href="#.-Specify-haplotypes-1" title="Permalink"></a></h3><p>Once you have a set of motifs defined, you can build a set of haplotypes by specifying sequences of motifs.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>You specify the motifs using a vector of ID numbers. If you use a negative ID number -N then it means the reverse complement of the sequence of motif N.</p></div></div><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>A motif&#39;s ID can be repeated in such a vector any number of times, so you can create repeat structures in a haplotype. </p></div></div><p>You add a haplotype by using the <code>add_motif_arrangement!</code> method with a <code>MotifStitcher</code> and a vector of motif IDs.</p><p>For example, let&#39;s make a sequence that would form a hair-pin like structure, with repeats when turned into a DeBruijn graph.</p><pre><code class="language-julia-repl">julia&gt; ms = MotifStitcher()
Pseudoseq.PuzzleMaker.MotifStitcher(0, Dict{Int64,Pseudoseq.PuzzleMaker.RandomMotif}(), Dict{Int64,BioSequences.LongSequence{BioSequences.DNAAlphabet{2}}}(), Dict{Int64,Pseudoseq.PuzzleMaker.SiblingMotif}(), Array{Int64,1}[])

julia&gt; add_motifs!(ms, 10000, 600, 10000, 600, 10000, 10000, 10000) # Use add_motifs! to add multiple random motifs at once.
Pseudoseq.PuzzleMaker.MotifStitcher(7, Dict(7 =&gt; Pseudoseq.PuzzleMaker.RandomMotif(10000, BioSequences.SamplerWeighted{BioSymbols.DNA}(BioSymbols.DNA[DNA_A, DNA_T, DNA_C, DNA_G], [0.25, 0.25, 0.25, 0.25])),4 =&gt; Pseudoseq.PuzzleMaker.RandomMotif(600, BioSequences.SamplerWeighted{BioSymbols.DNA}(BioSymbols.DNA[DNA_A, DNA_T, DNA_C, DNA_G], [0.25, 0.25, 0.25, 0.25])),2 =&gt; Pseudoseq.PuzzleMaker.RandomMotif(600, BioSequences.SamplerWeighted{BioSymbols.DNA}(BioSymbols.DNA[DNA_A, DNA_T, DNA_C, DNA_G], [0.25, 0.25, 0.25, 0.25])),3 =&gt; Pseudoseq.PuzzleMaker.RandomMotif(10000, BioSequences.SamplerWeighted{BioSymbols.DNA}(BioSymbols.DNA[DNA_A, DNA_T, DNA_C, DNA_G], [0.25, 0.25, 0.25, 0.25])),5 =&gt; Pseudoseq.PuzzleMaker.RandomMotif(10000, BioSequences.SamplerWeighted{BioSymbols.DNA}(BioSymbols.DNA[DNA_A, DNA_T, DNA_C, DNA_G], [0.25, 0.25, 0.25, 0.25])),6 =&gt; Pseudoseq.PuzzleMaker.RandomMotif(10000, BioSequences.SamplerWeighted{BioSymbols.DNA}(BioSymbols.DNA[DNA_A, DNA_T, DNA_C, DNA_G], [0.25, 0.25, 0.25, 0.25])),1 =&gt; Pseudoseq.PuzzleMaker.RandomMotif(10000, BioSequences.SamplerWeighted{BioSymbols.DNA}(BioSymbols.DNA[DNA_A, DNA_T, DNA_C, DNA_G], [0.25, 0.25, 0.25, 0.25]))), Dict{Int64,BioSequences.LongSequence{BioSequences.DNAAlphabet{2}}}(), Dict{Int64,Pseudoseq.PuzzleMaker.SiblingMotif}(), Array{Int64,1}[])

julia&gt; add_motif_arrangement!(ms, [1, 2, 3, 4, 5, -4, -6, -2, 7])
Pseudoseq.PuzzleMaker.MotifStitcher(7, Dict(7 =&gt; Pseudoseq.PuzzleMaker.RandomMotif(10000, BioSequences.SamplerWeighted{BioSymbols.DNA}(BioSymbols.DNA[DNA_A, DNA_T, DNA_C, DNA_G], [0.25, 0.25, 0.25, 0.25])),4 =&gt; Pseudoseq.PuzzleMaker.RandomMotif(600, BioSequences.SamplerWeighted{BioSymbols.DNA}(BioSymbols.DNA[DNA_A, DNA_T, DNA_C, DNA_G], [0.25, 0.25, 0.25, 0.25])),2 =&gt; Pseudoseq.PuzzleMaker.RandomMotif(600, BioSequences.SamplerWeighted{BioSymbols.DNA}(BioSymbols.DNA[DNA_A, DNA_T, DNA_C, DNA_G], [0.25, 0.25, 0.25, 0.25])),3 =&gt; Pseudoseq.PuzzleMaker.RandomMotif(10000, BioSequences.SamplerWeighted{BioSymbols.DNA}(BioSymbols.DNA[DNA_A, DNA_T, DNA_C, DNA_G], [0.25, 0.25, 0.25, 0.25])),5 =&gt; Pseudoseq.PuzzleMaker.RandomMotif(10000, BioSequences.SamplerWeighted{BioSymbols.DNA}(BioSymbols.DNA[DNA_A, DNA_T, DNA_C, DNA_G], [0.25, 0.25, 0.25, 0.25])),6 =&gt; Pseudoseq.PuzzleMaker.RandomMotif(10000, BioSequences.SamplerWeighted{BioSymbols.DNA}(BioSymbols.DNA[DNA_A, DNA_T, DNA_C, DNA_G], [0.25, 0.25, 0.25, 0.25])),1 =&gt; Pseudoseq.PuzzleMaker.RandomMotif(10000, BioSequences.SamplerWeighted{BioSymbols.DNA}(BioSymbols.DNA[DNA_A, DNA_T, DNA_C, DNA_G], [0.25, 0.25, 0.25, 0.25]))), Dict{Int64,BioSequences.LongSequence{BioSequences.DNAAlphabet{2}}}(), Dict{Int64,Pseudoseq.PuzzleMaker.SiblingMotif}(), Array{Int64,1}[[1, 2, 3, 4, 5, -4, -6, -2, 7]])</code></pre><h3 id=".-Generate-haplotype-sequences-1"><a class="docs-heading-anchor" href="#.-Generate-haplotype-sequences-1">3. Generate haplotype sequences</a><a class="docs-heading-anchor-permalink" href="#.-Generate-haplotype-sequences-1" title="Permalink"></a></h3><p>With the motifs defined and the haplotypes defined you can now generate sequences!</p><p>Simply call the <code>make_puzzle</code> method on the <code>MotifStitcher</code> to get a vector of haplotype sequences. Repeatedly call <code>make_puzzle</code> to get independently generated sequences from the same specification:</p><pre><code class="language-julia-repl">julia&gt; make_puzzle(ms)
1-element Array{BioSequences.LongSequence{BioSequences.DNAAlphabet{2}},1}:
 ACGACTAATCCTGGCGTGCCAATACTACTCAAACTGTAG…AGAGCCCCAGGAGGCTTACGCTGAGACATAGCCTACTAA

julia&gt; make_puzzle(ms)
1-element Array{BioSequences.LongSequence{BioSequences.DNAAlphabet{2}},1}:
 GCTCCGTCGTAATCTAATAACCGTCTGGCTAATGTTAGC…AGCGTGAAAGATATTGGGAGCTATGAGCATATGGATACA</code></pre><p>If you provide a filename, the haplotypes will be written to file in FASTA format, instead of returned as a value:</p><pre><code class="language-julia-repl">julia&gt; make_puzzle(ms, &quot;myhaplos.fasta&quot;)
1-element Array{BioSequences.LongSequence{BioSequences.DNAAlphabet{2}},1}:
 GGGGGCCTGCGGTGGAAGACAGGGGCAAAAGGTAGTCAG…CTGCAAGACTCTCTCGGCATAGGAGTGTGAAAGGGTCGA</code></pre><p>That&#39;s all there is to it. Now you can try a simulated sequencing experiment on your haplotypes. </p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../../../">« Home</a><a class="docs-footer-nextpage" href="../../sequencing/concepts/">Core concepts &amp; workflow »</a></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Sunday 24 May 2020 16:31">Sunday 24 May 2020</span>. Using Julia version 1.2.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
