<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Tagged paired end reads · Pseudoseq.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../../../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../../../../assets/documenter.js"></script><script src="../../../../siteinfo.js"></script><script src="../../../../../versions.js"></script><link href="../../../../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><a href="../../../../index.html"><img class="logo" src="../../../../assets/logo.png" alt="Pseudoseq.jl logo"/></a><h1>Pseudoseq.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../../../../search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../../../../">Home</a></li><li><span class="toctext">Manual</span><ul><li><span class="toctext">Sequencing</span><ul><li><a class="toctext" href="../../concepts/">Core concepts &amp; workflow</a></li><li><span class="toctext">Examples</span><ul><li><a class="toctext" href="../pe-example/">Paired end reads</a></li><li><a class="toctext" href="../se-example/">Long single end reads</a></li><li class="current"><a class="toctext" href>Tagged paired end reads</a><ul class="internal"><li><a class="toctext" href="#Using-the-[sequence](@ref)-method-1">Using the <code>sequence</code> method</a></li><li><a class="toctext" href="#Using-the-Pseudoseq-API-1">Using the Pseudoseq API</a></li></ul></li></ul></li></ul></li><li><span class="toctext">Build-a-Genome</span><ul><li><a class="toctext" href="../../../build-a-genome/concepts/">Core concepts &amp; workflow</a></li><li><span class="toctext">Examples</span><ul><li><a class="toctext" href="../../../build-a-genome/examples/build-a-yeast/">Build-a-Yeast</a></li></ul></li></ul></li></ul></li><li><span class="toctext">API</span><ul><li><span class="toctext">Sequencing</span><ul><li><a class="toctext" href="../../../../api/sequence/">sequence</a></li><li><a class="toctext" href="../../../../api/pool/">Molecule Pool</a></li><li><a class="toctext" href="../../../../api/reads/">Reads</a></li></ul></li><li><a class="toctext" href="../../../../api/chromosome-blueprint/">Build-a-Genome</a></li></ul></li></ul></nav><article id="docs"><header><nav><ul><li>Manual</li><li>Sequencing</li><li>Examples</li><li><a href>Tagged paired end reads</a></li></ul><a class="edit-page" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/master/examples/sequencing/tg-example.jl"><span class="fa"></span> Edit on GitHub</a></nav><hr/><div id="topbar"><span>Tagged paired end reads</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="Example:-tagged-paired-end-reads-1" href="#Example:-tagged-paired-end-reads-1">Example: tagged paired-end reads</a></h1><p>This is an example generated from this source file: <a href="https://github.com/bioinfologics/Pseudoseq.jl/blob/master/examples/sequencing/tg-example.jl"><code>tg-example.jl</code></a> You are seeing the online documentation version. The corresponding notebook can be found here: <a href="https://nbviewer.jupyter.org/github/bioinfologics/Pseudoseq.jl/blob/gh-pages/dev/examples/sequencing/tg-example.ipynb"><code>tg-example.ipynb</code></a> and the script can be found here: <a href="../tg-example.jl"><code>tg-example.jl</code></a></p><p>Let&#39;s see how we might simulate something like an 10x sequencing experiment.</p><p>For this simulation script we will:</p><ol><li>Create a pool of 5000 copies of a reference genome.</li><li>Fragment the DNA molecules in the pool, to an average length of 40,000bp.</li><li>Tag the long molecules in the pool randomly with a set of 1,000,000 tags.</li><li>Fragment the molecules in the pool to an average length of 700bp.</li><li>Subsample the molecules in the pool to achieve approximatly 50x coverage.</li><li>Create a set of 250bp paired-end reads.</li><li>Apply errors to the paired-end reads at a rate of 0.001 (.1%).</li><li>Generate an output FASTQ file.</li></ol><div><pre><code class="language-julia">using Pseudoseq</code></pre></div><h2><a class="nav-anchor" id="Using-the-[sequence](@ref)-method-1" href="#Using-the-[sequence](@ref)-method-1">Using the <a href="../../../../api/sequence/#Pseudoseq.sequence"><code>sequence</code></a> method</a></h2><p>First, let&#39;s see how we do this with the <code>sequence</code> method. The first two parameters we give to the function will be the input genome we want to sequence, and the destination FASTQ file for output reads. Here we are setting:</p><ul><li>The number of genome copies in the molecule pool to 5,000.</li><li>The number of possible tags to one million.</li><li>The average fragment size, prior to tagging, to 40,000bp.</li><li>The average fragment size after tagging, to 700bp.</li><li>The sampling coverage to 50x.</li><li>The read length to 250bp.</li><li>The per base read error rate to 0.001.</li><li>The fact we want paired-ends of fragments to be read (<code>paired</code>) to true, which is the default.</li></ul><div><pre><code class="language-julia">sequence(&quot;ecoli-ref.fasta&quot;, &quot;tagged_reads.fastq&quot;; ng = 5000, tusize = 1000000, taggedflen = 40000, flen = 700, cov = 50, rdlen = 250, err = 0.1)</code></pre><pre><code class="language-none">[ Info: - ✔ Created pool of 5000 copies of a 4639675bp genome
[ Info: - ✔ Created pool of fragments with an average length of 40000bp
[ Info: Attaching 1000000 tags randomly to each molecule in pool...
[ Info: Done
[ Info: - ✔ Created pool of tagged fragments with an average length of 40000bp
[ Info: - ✔ Created pool of fragments with an average length of 700bp
[ Info: - ✔ Subsampled pool at 50X coverage (463967 molecules)
[ Info: - ✔ Created set of 250bp paired-end reads
[ Info: - ✔ Applied sequencing errors at a per-base rate of 0.1
[ Info: - ✔ Wrote 654540 paired end reads to FASTQ file</code></pre></div><h2><a class="nav-anchor" id="Using-the-Pseudoseq-API-1" href="#Using-the-Pseudoseq-API-1">Using the Pseudoseq API</a></h2><p>Here&#39;s how to achieve the same thing, using the Pseudoseq API. It is nessecery to use the API if you want to do something that is not anticipated by the available functionality of the <code>sequence</code> method: the cost of conveinience is fewer options.</p><p>Let&#39;s start with a pool of 5000 copies of a genome contained in a FASTA file:</p><div><pre><code class="language-julia">dnapool = makepool(&quot;ecoli-ref.fasta&quot;, 5000)</code></pre><pre><code class="language-none">Pool of 5000 molecules:
 All molecules are of the same size: 4639675</code></pre></div><p>Now let&#39;s cut up the molecules to an average length of 40,000bp</p><div><pre><code class="language-julia">cutpool = fragment(dnapool, 40000)</code></pre><pre><code class="language-none">Pool of 575000 molecules:
 Maximum molecule size: 511813
 Average molecule size: 40345
 Minimum molecule size: 1</code></pre></div><p>Ok, now we will tag these large fragments randomly. Once you tag a fragment in a universe, any other fragments that are derived from that tagged fragment will inherit the same tag.</p><div><pre><code class="language-julia">taggedpool = tag(cutpool, 1000000)</code></pre><pre><code class="language-none">[ Info: Attaching 1000000 tags randomly to each molecule in pool...
[ Info: Done
Pool of 575000 molecules:
 Maximum molecule size: 511813
 Average molecule size: 40345
 Minimum molecule size: 1
 Number of distinct tags: 437499</code></pre></div><p>Here I&#39;m going to use a pool of 1,000,000 distinct tags. Which fragment gets a certain tag is random. The size of the tag pool, and the number of fragments in your universe will determine how likely it is that any two fragments get the same tag. Now we&#39;ll fragment the pool again</p><div><pre><code class="language-julia">taggedcutpool = fragment(taggedpool, 700)</code></pre><pre><code class="language-none">Pool of 32863162 molecules:
 Maximum molecule size: 12206
 Average molecule size: 705
 Minimum molecule size: 1
 Number of distinct tags: 437499</code></pre></div><p>Subsample the pool of tagged molecules.</p><div><pre><code class="language-julia">genome_size = 4639675
expected_coverage = 50
read_length = 250

N = needed_sample_size(expected_coverage, genome_size, read_length)
N = div(N, 2) # Divide by 2 as we&#39;re doing paired end sequencing.

sampledpool = subsample(taggedcutpool, N)</code></pre><pre><code class="language-none">Pool of 463967 molecules:
 Maximum molecule size: 9661
 Average molecule size: 704
 Minimum molecule size: 1
 Number of distinct tags: 226652</code></pre></div><p>Now let&#39;s make some 250bp tagged paired reads and generate some erroneous positions.</p><div><pre><code class="language-julia">tagged_reads = make_reads(TaggedPairs, sampledpool, 250)
tagged_w_errs = mark_errors(tagged_reads, 0.001)</code></pre></div><p>Output to FASTQ:</p><div><pre><code class="language-julia">generate(&quot;tagged_reads.fastq&quot;, tagged_w_errs)#-</code></pre><pre><code class="language-none">[ Info: - ✔ Wrote 654390 tagged paired end reads to FASTQ file</code></pre></div><p><em>This page was generated using <a href="https://github.com/fredrikekre/Literate.jl">Literate.jl</a>.</em></p><footer><hr/><a class="previous" href="../se-example/"><span class="direction">Previous</span><span class="title">Long single end reads</span></a><a class="next" href="../../../build-a-genome/concepts/"><span class="direction">Next</span><span class="title">Core concepts &amp; workflow</span></a></footer></article></body></html>
