<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Tagged paired end reads · Pseudoseq.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link href="../../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><a href="../../index.html"><img class="logo" src="../../assets/logo.png" alt="Pseudoseq.jl logo"/></a><h1>Pseudoseq.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../../search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../../">Home</a></li><li><span class="toctext">User manual</span><ul><li><a class="toctext" href="../introduction/">Introduction</a></li><li><span class="toctext">Walkthroughs</span><ul><li><a class="toctext" href="../pe_walkthrough/">Paired end reads</a></li><li><a class="toctext" href="../se_walkthrough/">Long single end reads</a></li><li class="current"><a class="toctext" href>Tagged paired end reads</a><ul class="internal"><li><a class="toctext" href="#Creating-the-universe-1">Creating the universe</a></li><li><a class="toctext" href="#Fragmenting-the-molecules-1">Fragmenting the molecules</a></li><li><a class="toctext" href="#Tag-the-large-fragments-1">Tag the large fragments</a></li><li><a class="toctext" href="#Fragmenting-the-molecules,-again-1">Fragmenting the molecules, again</a></li><li><a class="toctext" href="#Subsampling-the-molecules-1">Subsampling the molecules</a></li><li><a class="toctext" href="#Creating-the-reads-1">Creating the reads</a></li><li><a class="toctext" href="#Generate-FASTQ-files-1">Generate FASTQ files</a></li></ul></li></ul></li><li><span class="toctext">API</span><ul><li><a class="toctext" href="../../api/universe/">Universe</a></li></ul></li></ul></li></ul></nav><article id="docs"><header><nav><ul><li>User manual</li><li>Walkthroughs</li><li><a href>Tagged paired end reads</a></li></ul><a class="edit-page" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/master/docs/src/man/tg_walkthrough.md"><span class="fa"></span> Edit on GitHub</a></nav><hr/><div id="topbar"><span>Tagged paired end reads</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="Walkthrough:-tagged-paired-end-reads-1" href="#Walkthrough:-tagged-paired-end-reads-1">Walkthrough: tagged paired end reads</a></h1><p>Let&#39;s see how you might simulate something like an 10x sequencing experiment.</p><p>This experiment is simulated much in the same way as demonstrated in the paired-end sequencing walkthrough, except that there is an extra fragmentation step and a tagging step. Let&#39;s begin!</p><h2><a class="nav-anchor" id="Creating-the-universe-1" href="#Creating-the-universe-1">Creating the universe</a></h2><p>Let&#39;s create a universe like the one in the paired-end sequencing walkthrough.</p><pre><code class="language-julia">universe = makeuniverse(&quot;mygenome.fasta&quot;, 5000)</code></pre><h2><a class="nav-anchor" id="Fragmenting-the-molecules-1" href="#Fragmenting-the-molecules-1">Fragmenting the molecules</a></h2><p>Ok, this time we are going to fragment the molecules, but first at a larger size:</p><pre><code class="language-julia">big_cut_universe = fragment(universe, 40000)</code></pre><p>Now we have a universe of fragments approximately 40,000bp in size.</p><h2><a class="nav-anchor" id="Tag-the-large-fragments-1" href="#Tag-the-large-fragments-1">Tag the large fragments</a></h2><p>Ok, now we will tag these large fragments randomly, using a pool of N distinct tags. Once you tag a fragment in a universe, any other fragments that are derived from that tagged fragment will inherit the same tag. For example, if a DNA fragment in a universe is tagged, and then it is subsequently fragmented during a <code>fragment</code> transform, then all the smaller fragments derived from that long fragment will inherit that long fragments tag. This allows shorter fragments to possess longer range information in the form of these tags, and this is the basis of 10x and similar technologies.</p><pre><code class="language-julia">tagged_universe = tag(big_cut_universe, 1000000)</code></pre><p>Here I&#39;m going to use a pool of 1,000,000 distinct tags. Which fragment gets a certain tag is random. The size of the tag pool, and the number of fragments in your universe will determine how likely it is that any two fragments get the same tag.</p><h2><a class="nav-anchor" id="Fragmenting-the-molecules,-again-1" href="#Fragmenting-the-molecules,-again-1">Fragmenting the molecules, again</a></h2><p>Ok now this time we will fragment the large tagged molecules in our universe down to an average expected fragment size of 700bp, just as we did when we were simulating paired-end sequencing.</p><pre><code class="language-julia">cut_universe = fragment(tagged_universe, 700)</code></pre><h2><a class="nav-anchor" id="Subsampling-the-molecules-1" href="#Subsampling-the-molecules-1">Subsampling the molecules</a></h2><p>Ok, now let&#39;s subsample the universe just as we did for the paired end sequencing simulation, since the genome size, and read length is the same as in that simulation. We already know to sample 463967 fragments in order to achieve approximately 50x coverage:</p><pre><code class="language-julia">subsampled_universe = subsample(cut_universe, 463967)</code></pre><h2><a class="nav-anchor" id="Creating-the-reads-1" href="#Creating-the-reads-1">Creating the reads</a></h2><p>We now want to create a set of paired-end reads. We want our reads to be 250bp in length.</p><p>We use the <code>make_reads</code> function to achieve this:</p><pre><code class="language-julia">tagged_reads = make_reads(TaggedPairs, subsampled_universe, 250)</code></pre><p>Now we have some reads, we should mark positions in the reads that are destined to be errors in the output FASTQ.</p><p>We do this using the <code>mark_errors</code> function. The function requires an error rate.</p><pre><code class="language-julia">tagged_w_errs = mark_errors(tagged_reads, 0.001)</code></pre><p>Currently, every position in every read is equally likely to be marked as an error.</p><h2><a class="nav-anchor" id="Generate-FASTQ-files-1" href="#Generate-FASTQ-files-1">Generate FASTQ files</a></h2><p>Now we have some paired end reads and have marked some positions as errors, we can generate FASTQ files. This is done with the <code>generate</code> function.</p><pre><code class="language-julia">generate(&quot;tagged_reads.fastq&quot;, tagged_w_errs)</code></pre><footer><hr/><a class="previous" href="../se_walkthrough/"><span class="direction">Previous</span><span class="title">Long single end reads</span></a><a class="next" href="../../api/universe/"><span class="direction">Next</span><span class="title">Universe</span></a></footer></article></body></html>
