<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Paired end reads · Pseudoseq.jl</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../../../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../../../assets/documenter.js"></script><script src="../../../../siteinfo.js"></script><script src="../../../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../../../"><img src="../../../../assets/logo.png" alt="Pseudoseq.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit">Pseudoseq.jl</span></div><form class="docs-search" action="../../../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../../../">Home</a></li><li><span class="tocitem">Manual</span><ul><li><a class="tocitem" href="../../../puzzlemaker/concepts/">PuzzleMaker</a></li><li><input class="collapse-toggle" id="menuitem-2-2" type="checkbox" checked/><label class="tocitem" for="menuitem-2-2"><span class="docs-label">Sequencing</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../concepts/">Core concepts &amp; workflow</a></li><li><input class="collapse-toggle" id="menuitem-2-2-2" type="checkbox" checked/><label class="tocitem" for="menuitem-2-2-2"><span class="docs-label">Examples</span><i class="docs-chevron"></i></label><ul class="collapsed"><li class="is-active"><a class="tocitem" href>Paired end reads</a></li><li><a class="tocitem" href="../se-example/">Long single end reads</a></li><li><a class="tocitem" href="../tg-example/">Tagged paired end reads</a></li></ul></li></ul></li></ul></li><li><span class="tocitem">API</span><ul><li><input class="collapse-toggle" id="menuitem-3-1" type="checkbox"/><label class="tocitem" for="menuitem-3-1"><span class="docs-label">Sequencing</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../../../api/sequence/">sequence</a></li><li><a class="tocitem" href="../../../../api/pool/">Molecule Pool</a></li><li><a class="tocitem" href="../../../../api/reads/">Reads</a></li></ul></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Manual</a></li><li><a class="is-disabled">Sequencing</a></li><li><a class="is-disabled">Examples</a></li><li class="is-active"><a href>Paired end reads</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Paired end reads</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/master/examples/sequencing/pe-example.jl" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Example:-paired-end-sequencing-1"><a class="docs-heading-anchor" href="#Example:-paired-end-sequencing-1">Example: paired-end sequencing</a><a class="docs-heading-anchor-permalink" href="#Example:-paired-end-sequencing-1" title="Permalink"></a></h1><p>This is an example generated from this source file: <a href="https://github.com/bioinfologics/Pseudoseq.jl/blob/master/examples/sequencing/pe-example.jl"><code>pe-example.jl</code></a> You are seeing the online documentation version. The corresponding notebook can be found here: <a href="https://nbviewer.jupyter.org/github/bioinfologics/Pseudoseq.jl/blob/gh-pages/dev/man/sequencing/examples/pe-example.ipynb"><code>pe-example.ipynb</code></a> and the script can be found here: <a href="../pe-example.jl"><code>pe-example.jl</code></a></p><p>For the simulation we are going to:</p><ol><li>Create a pool of 5000 copies of a reference genome.</li><li>Fragment the DNA molecules in the pool, to an average length of 700bp.</li><li>Subsample the molecules in the pool to achieve approximatly 50x coverage.</li><li>Create a set of 250bp paired-end reads.</li><li>Apply errors to the paired-end reads at a rate of 0.001 (.1%).</li><li>Generate an output FASTQ file.</li></ol><pre><code class="language-julia">using Pseudoseq.Sequencing</code></pre><p>Starting with a FASTA formatted file containing the genome we want to sequence, we create a pool with a copy of the genome.</p><pre><code class="language-julia">m = Molecules(&quot;ecoli-ref.fasta&quot;)</code></pre><pre><code class="language-none">Pool of 1 DNA molecules:
 All molecules are of the same size: 4639675
</code></pre><p>We want to first have 5000 full copies of the genome, so we will make an amplifier.</p><pre><code class="language-julia">amp = amplify(5000)</code></pre><pre><code class="language-none">Pseudoseq.Sequencing.Amplifier{Pseudoseq.Sequencing.var&quot;#5#6&quot;}(Pseudoseq.Sequencing.var&quot;#5#6&quot;(), 5000)</code></pre><p>Next we want a fragmenter to make a pool of shorter DNA molecules, with an average length of 700bp.</p><pre><code class="language-julia">frag = fragment(700bp)</code></pre><pre><code class="language-none">Pseudoseq.Sequencing.Fragmenter(700)</code></pre><p>We will want a filter to make sure no extreme sized molecules make their way through.</p><pre><code class="language-julia">size_filter = select(x -&gt; 900 &gt;= length(x) &gt;= 450)</code></pre><pre><code class="language-none">Pseudoseq.Sequencing.Selector{Main.ex-pe-example.var&quot;#1#2&quot;}(Main.ex-pe-example.var&quot;#1#2&quot;())</code></pre><p>Next we create a subsampler which will randomly sample molecules, to give us a desired expected coverage. We will want our reads to be 250bp in length.</p><pre><code class="language-julia">ssmpl = subsample(50X, 2 * 250bp)</code></pre><pre><code class="language-none">Pseudoseq.Sequencing.CovSubSampler{Tuple{Pseudoseq.Sequencing.SequenceLength,Pseudoseq.Sequencing.SequenceLength}}(50X sequencing coverage
, (250 base pairs
, 250 base pairs
))</code></pre><p>We then want a read-maker that will give us paired end reads. We want our reads to be 250bp in length.</p><pre><code class="language-julia">readmaker = makereads(2 * 250bp)</code></pre><pre><code class="language-none">Pseudoseq.Sequencing.PairedReads(250, 250)</code></pre><p>Once we have reads, we will mark positions in the reads that are incorrectly detected by the sequencer: errors. We will construct a <code>FixedProbSubstitutions</code> function with a per base error probability of 0.001 and pass it to the <code>make_substitutions</code> method. This will make errors fall totally randomly over each read.</p><pre><code class="language-julia">errmaker = make_substitutions(FixedProbSubstitutions(0.001))</code></pre><pre><code class="language-none">Pseudoseq.Sequencing.SubstitutionMaker{FixedProbSubstitutions}(FixedProbSubstitutions(0.001))</code></pre><p>Now we can push our molecules through a pipeline of these processors, and out to a FASTQ file:</p><pre><code class="language-julia">m |&gt; amp |&gt; frag |&gt; size_filter |&gt; ssmpl |&gt; readmaker |&gt; errmaker |&gt; generate(&quot;pe-reads.fastq&quot;)</code></pre><pre><code class="language-none"></code></pre><p>You can also compose the processors together into one whole function. Typing \circ in the julia repl and then hitting tab gives you the circular composition symbol. Note how pipelining above progresses from left to right, but composition is right to left in order.</p><pre><code class="language-julia">my_protocol = errmaker ∘ readmaker ∘ ssmpl ∘ size_filter ∘ frag ∘ amp

m |&gt; my_protocol |&gt; generate(&quot;pe-reads.fastq&quot;)</code></pre><pre><code class="language-none"></code></pre><hr/><p><em>This page was generated using <a href="https://github.com/fredrikekre/Literate.jl">Literate.jl</a>.</em></p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../../concepts/">« Core concepts &amp; workflow</a><a class="docs-footer-nextpage" href="../se-example/">Long single end reads »</a></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Thursday 24 September 2020 00:29">Thursday 24 September 2020</span>. Using Julia version 1.5.1.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
