<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Long single end reads · Pseudoseq.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link href="../../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><a href="../../index.html"><img class="logo" src="../../assets/logo.png" alt="Pseudoseq.jl logo"/></a><h1>Pseudoseq.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../../search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../../">Home</a></li><li><span class="toctext">User manual</span><ul><li><a class="toctext" href="../introduction/">Introduction</a></li><li><span class="toctext">Walkthroughs</span><ul><li><a class="toctext" href="../pe_walkthrough/">Paired end reads</a></li><li class="current"><a class="toctext" href>Long single end reads</a><ul class="internal"><li><a class="toctext" href="#Create-a-universe-1">Create a universe</a></li><li><a class="toctext" href="#Fragmenting-the-molecules-1">Fragmenting the molecules</a></li><li><a class="toctext" href="#Creating-the-reads-1">Creating the reads</a></li></ul></li><li><a class="toctext" href="../tg_walkthrough/">Tagged paired end reads</a></li></ul></li><li><span class="toctext">API</span><ul><li><a class="toctext" href="../../api/universe/">Universe</a></li><li><a class="toctext" href="../../api/reads/">Reads</a></li></ul></li></ul></li></ul></nav><article id="docs"><header><nav><ul><li>User manual</li><li>Walkthroughs</li><li><a href>Long single end reads</a></li></ul><a class="edit-page" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/master/docs/src/man/se_walkthrough.md"><span class="fa"></span> Edit on GitHub</a></nav><hr/><div id="topbar"><span>Long single end reads</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="Walkthrough:-Long,-single-end-reads-1" href="#Walkthrough:-Long,-single-end-reads-1">Walkthrough: Long, single end reads</a></h1><p>Let&#39;s see how you might simulate something like an Oxford Nanopore sequencing experiment.</p><h2><a class="nav-anchor" id="Create-a-universe-1" href="#Create-a-universe-1">Create a universe</a></h2><p>Let&#39;s start with a universe just as we did for the paired-end reads walkthrough:</p><pre><code class="language-julia">universe = makeuniverse(&quot;mygenome.fasta&quot;, 5000)</code></pre><h2><a class="nav-anchor" id="Fragmenting-the-molecules-1" href="#Fragmenting-the-molecules-1">Fragmenting the molecules</a></h2><p>Now we <a href="../../api/universe/#Pseudoseq.fragment-Tuple{Pseudoseq.Universe,Int64}"><code>fragment</code></a> the molecules in our universe, but this time we will use a much larger fragment size, to simulate higher molecular weight DNA that long read technologies typically require:</p><pre><code class="language-julia">cut_universe = fragment(universe, 40000)</code></pre><p>Now we have a universe of large cut fragments. We can do a subsample of the universe. We will estimate the number of fragments to sample using the average length of the molecules (40000). We will use the <code>needed_sample_size</code> function to help us out, as we did in the paired end walkthrough.</p><pre><code class="language-julia">genome_size = 4639675
expected_coverage = 30
readlength = 40000

N = needed_sample_size(expected_coverage, genome_size, readlength)

uni_sample = subsample(cut_universe, N)</code></pre><h2><a class="nav-anchor" id="Creating-the-reads-1" href="#Creating-the-reads-1">Creating the reads</a></h2><p>Ok now we&#39;ve come to creating the reads, we do this with the <a href="../../api/reads/#Pseudoseq.make_reads"><code>make_reads</code></a> function. By only specifying the read type, and the sample of our universe, and omitting the read length we want. Make reads will generate reads from the entire length of each molecule in our sample. We elected to do this to emulate what Nanopore sequencing is supposed to do: It takes an entire DNA fragment, feeds it through an electrically charged pore, producing a read for the entire fragment. </p><pre><code class="language-julia">se_reads = make_reads(SingleEnd, uni_sample)</code></pre><p>Now we have some reads, we should mark positions in the reads that are destined to be errors in the output FASTQ.</p><p>We do this using the <a href="../../api/reads/#Pseudoseq.mark_errors"><code>mark_errors</code></a> function. The function requires an error rate.</p><pre><code class="language-julia">se_w_errs = mark_errors(se_reads, 0.1)</code></pre><p>We&#39;ve marked many more positions as errors than we did for the paired end tutorial, this is because current long read sequencing technology has a much higher error rate and short read sequencing technologies, in general.</p><p>Ok, now we have marked the errors, we can <a href="../../api/reads/#Pseudoseq.generate"><code>generate</code></a> a FASTQ file with these reads:</p><pre><code class="language-julia">generate(&quot;longreads.fastq&quot;, se_w_errs)</code></pre><p>And thats all there is to it!</p><footer><hr/><a class="previous" href="../pe_walkthrough/"><span class="direction">Previous</span><span class="title">Paired end reads</span></a><a class="next" href="../tg_walkthrough/"><span class="direction">Next</span><span class="title">Tagged paired end reads</span></a></footer></article></body></html>
