<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Sequencing · Pseudoseq.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link href="../../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><a href="../../index.html"><img class="logo" src="../../assets/logo.png" alt="Pseudoseq.jl logo"/></a><h1>Pseudoseq.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../../search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../../">Home</a></li><li><span class="toctext">Build-a-Genome</span><ul><li><a class="toctext" href="../../build-a-genome/">Core concepts &amp; workflow</a></li><li><span class="toctext">Examples</span></li></ul></li><li><span class="toctext">Sequencing</span><ul><li><a class="toctext" href="../../sequencing/">Core concepts &amp; workflow</a></li><li><span class="toctext">Examples</span><ul><li><a class="toctext" href="../../examples/pe-example/">Paired end reads</a></li><li><a class="toctext" href="../../examples/se-example/">Long single end reads</a></li><li><a class="toctext" href="../../examples/tg-example/">Tagged paired end reads</a></li></ul></li></ul></li><li><span class="toctext">API</span><ul><li class="current"><a class="toctext" href>Sequencing</a><ul class="internal"><li><a class="toctext" href="#Exported-functions-1">Exported functions</a></li></ul></li><li><a class="toctext" href="../pool/">Molecule Pool</a></li><li><a class="toctext" href="../reads/">Reads</a></li><li><a class="toctext" href="../chromosome-blueprint/">Chromosome Blueprint</a></li></ul></li></ul></nav><article id="docs"><header><nav><ul><li>API</li><li><a href>Sequencing</a></li></ul><a class="edit-page" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/master/docs/src/api/sequence.md"><span class="fa"></span> Edit on GitHub</a></nav><hr/><div id="topbar"><span>Sequencing</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="API:-sequence-1" href="#API:-sequence-1">API: sequence</a></h1><h2><a class="nav-anchor" id="Exported-functions-1" href="#Exported-functions-1">Exported functions</a></h2><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Pseudoseq.sequence" href="#Pseudoseq.sequence"><code>Pseudoseq.sequence</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">sequence(input, output = nothing; kwargs...)</code></pre><p>Run a sequencing experiment from start to finish.</p><p>This method is the highest-level function for running sequencing simulations with Pseudoseq. It runs the Pseudoseq sequencin worksflow described in the manual, tuned by the following keyword parameters:</p><p>Global parameters:</p><ul><li><code>input</code>: A filename or <code>FASTA.Reader</code> providing the input genome.</li><li><code>output</code>: A filename or <code>FASTA.Writer</code> providing a destination for the output reads.</li></ul><p>Parameters for workflow step 1 (Create a pool of DNA molecules):</p><ul><li><code>ng</code>: Integer; the number of genomes the molecule pool is initialized with.</li></ul><p>Parameters for step 2 (Processing the DNA molecule pool):</p><ul><li><code>flen</code>: Integer; the desired average fragment length (default: 700bp).</li><li><code>cov</code>: Integer; the desired expected coverage (default: 30x coverage).</li></ul><p>Parameters for step 3 (Generating reads):</p><ul><li><code>paired</code>: true or false; Whether or not to sequence from both end of each molecule (default: true).</li><li><code>rdlen</code>: Integer; the desired read length (default: 250bp).</li><li><code>err</code>: Float; the desired per-base error rate (default: 0.001).</li></ul></div></div><a class="source-link" target="_blank" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/8220254462975893648580df4b486ce7907ddb11/src/sequence.jl#L1-L26">source</a></section><footer><hr/><a class="previous" href="../../examples/tg-example/"><span class="direction">Previous</span><span class="title">Tagged paired end reads</span></a><a class="next" href="../pool/"><span class="direction">Next</span><span class="title">Molecule Pool</span></a></footer></article></body></html>
