<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Reads · Pseudoseq.jl</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img src="../../assets/logo.png" alt="Pseudoseq.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit">Pseudoseq.jl</span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Manual</span><ul><li><a class="tocitem" href="../../man/puzzlemaker/concepts/">PuzzleMaker</a></li><li><input class="collapse-toggle" id="menuitem-2-2" type="checkbox"/><label class="tocitem" for="menuitem-2-2"><span class="docs-label">Sequencing</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../man/sequencing/concepts/">Core concepts &amp; workflow</a></li><li><input class="collapse-toggle" id="menuitem-2-2-2" type="checkbox"/><label class="tocitem" for="menuitem-2-2-2"><span class="docs-label">Examples</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../man/sequencing/examples/pe-example/">Paired end reads</a></li><li><a class="tocitem" href="../../man/sequencing/examples/se-example/">Long single end reads</a></li><li><a class="tocitem" href="../../man/sequencing/examples/tg-example/">Tagged paired end reads</a></li></ul></li></ul></li></ul></li><li><span class="tocitem">API</span><ul><li><input class="collapse-toggle" id="menuitem-3-1" type="checkbox" checked/><label class="tocitem" for="menuitem-3-1"><span class="docs-label">Sequencing</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../sequence/">sequence</a></li><li><a class="tocitem" href="../pool/">Molecule Pool</a></li><li class="is-active"><a class="tocitem" href>Reads</a><ul class="internal"><li><a class="tocitem" href="#Exported-functions-1"><span>Exported functions</span></a></li></ul></li></ul></li><li><a class="tocitem" href="../chromosome-blueprint/">Build-a-Genome</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">API</a></li><li><a class="is-disabled">Sequencing</a></li><li class="is-active"><a href>Reads</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Reads</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/master/docs/src/api/reads.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="API:-Reads-1"><a class="docs-heading-anchor" href="#API:-Reads-1">API: Reads</a><a class="docs-heading-anchor-permalink" href="#API:-Reads-1" title="Permalink"></a></h1><h2 id="Exported-functions-1"><a class="docs-heading-anchor" href="#Exported-functions-1">Exported functions</a><a class="docs-heading-anchor-permalink" href="#Exported-functions-1" title="Permalink"></a></h2><h3 id="Making-reads-1"><a class="docs-heading-anchor" href="#Making-reads-1">Making reads</a><a class="docs-heading-anchor-permalink" href="#Making-reads-1" title="Permalink"></a></h3><article class="docstring"><header><a class="docstring-binding" id="Pseudoseq.Sequencing.paired_reads" href="#Pseudoseq.Sequencing.paired_reads"><code>Pseudoseq.Sequencing.paired_reads</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">paired_reads(p::MoleculePool, flen::Int, rlen::Int = flen)</code></pre><p>Create a set of paired-end reads from a pool of DNA molecules <code>p</code>.</p><p><code>flen</code> sets the length of forward read, and <code>rlen</code> sets the length of the reverse read. If you only provide <code>flen</code>, then the function sets <code>rlen = flen</code>.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>If a molecule in the pool is not long enough to create a forward and/or reverse read, then that molecule will simply be skipped. </p></div></div></div></section></article><article class="docstring"><header><a class="docstring-binding" id="Pseudoseq.Sequencing.unpaired_reads" href="#Pseudoseq.Sequencing.unpaired_reads"><code>Pseudoseq.Sequencing.unpaired_reads</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">unpaired_reads(p::Molecules{T}, len::Int) where {T&lt;:AbstractSequencingView}</code></pre><p>Create a set of single-end reads from a pool of DNA molecules <code>p</code>.</p><p><code>len</code> sets the length of the reads.</p><p>The end (strand) from which the reading begins for each DNA molecule in the pool is determined at random for each molecule, with 50:50 probability.</p><p>If you don&#39;t provide a value for <code>len</code>, then the function will read each DNA molecule in it&#39;s entirety.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>If a molecule in the pool is not long enough to create a forward and/or reverse read, then that molecule will simply be skipped. </p></div></div></div></section></article><h3 id="Introducing-errors-1"><a class="docs-heading-anchor" href="#Introducing-errors-1">Introducing errors</a><a class="docs-heading-anchor-permalink" href="#Introducing-errors-1" title="Permalink"></a></h3><article class="docstring"><header><a class="docstring-binding" id="Pseudoseq.Sequencing.edit_substitutions" href="#Pseudoseq.Sequencing.edit_substitutions"><code>Pseudoseq.Sequencing.edit_substitutions</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">edit_substitutions(f::Function, reads::Reads)</code></pre><p>Add or remove sequencing errors to reads in the form of single base substitutions, according to so some error generating or deleting function <code>f</code>.</p><p>The function provided must be a function that accepts two arguments:     1. A vector of Substitution structs, which the function will mutate.     2. The nucleotide sequence of the read.</p><p>The function needs not return anything, and anything it does return will not be used.</p><p>Some acceptable functions already provided with Pseudoseq include <a href="#Pseudoseq.Sequencing.FixedProbSubstitutions"><code>FixedProbSubstitutions</code></a> and <a href="#Pseudoseq.Sequencing.ClearSubstitutions"><code>ClearSubstitutions</code></a>. If you want to develop your own function, their implementations provide a simple example of how it can be done. You can of course also just create an anonymous function.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="Pseudoseq.Sequencing.FixedProbSubstitutions" href="#Pseudoseq.Sequencing.FixedProbSubstitutions"><code>Pseudoseq.Sequencing.FixedProbSubstitutions</code></a> — <span class="docstring-category">Type</span></header><section><div><p>A simple function object that can be used with <a href="#Pseudoseq.Sequencing.edit_substitutions"><code>edit_substitutions</code></a> or <a href="api/@ref"><code>edit_substitutions!</code></a>.</p><p>Randomly applies sequencing errors - substitutions - to all reads according to a fixed, uniform, per-base probability.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="Pseudoseq.Sequencing.ClearSubstitutions" href="#Pseudoseq.Sequencing.ClearSubstitutions"><code>Pseudoseq.Sequencing.ClearSubstitutions</code></a> — <span class="docstring-category">Type</span></header><section><div><p>A simple function object that can be used with <a href="#Pseudoseq.Sequencing.edit_substitutions"><code>edit_substitutions</code></a> or <a href="api/@ref"><code>edit_substitutions!</code></a>.</p><p>Simply clears all sequencing errors - substitutions - for all reads.</p></div></section></article><h3 id="Generating-FASTQ-files-1"><a class="docs-heading-anchor" href="#Generating-FASTQ-files-1">Generating FASTQ files</a><a class="docs-heading-anchor-permalink" href="#Generating-FASTQ-files-1" title="Permalink"></a></h3><article class="docstring"><header><a class="docstring-binding" id="Pseudoseq.Sequencing.generate" href="#Pseudoseq.Sequencing.generate"><code>Pseudoseq.Sequencing.generate</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">generate(R1name::String, R2name::String, reads::Reads{Paired,&lt;:AbstractSequencingView})</code></pre><p>This method only works for paired reads. Instead of interleaving R1 and R2 reads in a single FASTQ file, R1 and R2 reads are partitioned into two seperate FASTQ files.</p></div></section><section><div><pre><code class="language-none">generate(filename::String, reads::Reads)</code></pre><p>Write the <code>reads</code> out to a FASTQ formatted file with the given <code>filename</code>.</p><p>If this method is used with a paired-end read type, then the FASTQ file will be interleaved; all R1 reads will be odd records, and all R2 reads will be even records in the file.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>Reads are named according to the sequence in the input genome they came from. e.g. <code>@Reference_1_R1</code> means the first sequence in the genome, and <code>@Reference_2_R1</code> means the second sequence in the genome.</p></div></div></div></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../pool/">« Molecule Pool</a><a class="docs-footer-nextpage" href="../chromosome-blueprint/">Build-a-Genome »</a></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Monday 18 May 2020 11:34">Monday 18 May 2020</span>. Using Julia version 1.2.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
