<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Long single end reads · Pseudoseq.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../../../assets/documenter.js"></script><script src="../../../siteinfo.js"></script><script src="../../../../versions.js"></script><link href="../../../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><a href="../../../index.html"><img class="logo" src="../../../assets/logo.png" alt="Pseudoseq.jl logo"/></a><h1>Pseudoseq.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../../../search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../../../">Home</a></li><li><span class="toctext">Sequencing</span><ul><li><a class="toctext" href="../../../sequencing/">Core concepts &amp; workflow</a></li><li><span class="toctext">Examples</span><ul><li><a class="toctext" href="../pe-example/">Paired end reads</a></li><li class="current"><a class="toctext" href>Long single end reads</a><ul class="internal"><li><a class="toctext" href="#Using-the-[sequence](@ref)-method-1">Using the <code>sequence</code> method</a></li><li><a class="toctext" href="#Using-the-Pseudoseq-API-1">Using the Pseudoseq API</a></li></ul></li><li><a class="toctext" href="../tg-example/">Tagged paired end reads</a></li></ul></li></ul></li><li><span class="toctext">Build-a-Genome</span><ul><li><a class="toctext" href="../../../build-a-genome/">Core concepts &amp; workflow</a></li><li><span class="toctext">Examples</span><ul><li><a class="toctext" href="../../build-a-genome/build-a-yeast/">Build-a-Yeast</a></li></ul></li></ul></li><li><span class="toctext">API</span><ul><li><span class="toctext">Sequencing</span><ul><li><a class="toctext" href="../../../api/sequence/">sequence</a></li><li><a class="toctext" href="../../../api/pool/">Molecule Pool</a></li><li><a class="toctext" href="../../../api/reads/">Reads</a></li></ul></li><li><a class="toctext" href="../../../api/chromosome-blueprint/">Build-a-Genome</a></li></ul></li></ul></nav><article id="docs"><header><nav><ul><li>Sequencing</li><li>Examples</li><li><a href>Long single end reads</a></li></ul><a class="edit-page" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/master/examples/sequencing/se-example.jl"><span class="fa"></span> Edit on GitHub</a></nav><hr/><div id="topbar"><span>Long single end reads</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="Example:-long,-single-end-reads-1" href="#Example:-long,-single-end-reads-1">Example: long, single end reads</a></h1><p>This is an example generated from this source file: <a href="https://github.com/bioinfologics/Pseudoseq.jl/blob/master/examples/sequencing/se-example.jl"><code>se-example.jl</code></a> You are seeing the online documentation version. The corresponding notebook can be found here: <a href="https://nbviewer.jupyter.org/github/bioinfologics/Pseudoseq.jl/blob/gh-pages/dev/examples/sequencing/se-example.ipynb"><code>se-example.ipynb</code></a> and the script can be found here: <a href="../se-example.jl"><code>se-example.jl</code></a></p><p>Let&#39;s see how you might simulate something like an Oxford Nanopore sequencing experiment.</p><p>For the simulation we are going to:</p><ol><li>Create a pool of 5000 copies of a reference genome.</li><li>Fragment the DNA molecules in the pool, to an average length of 40,000bp.</li><li>Subsample the molecules in the pool to achieve approximatly 30x coverage.</li><li>Create a set of single-end reads, the enitre length of each molecule.</li><li>Apply errors to the reads at a rate of 0.10 (1 error every 10bp).</li><li>Generate an output FASTQ file.</li></ol><div><pre><code class="language-julia">using Pseudoseq</code></pre></div><h2><a class="nav-anchor" id="Using-the-[sequence](@ref)-method-1" href="#Using-the-[sequence](@ref)-method-1">Using the <a href="../../../api/sequence/#Pseudoseq.sequence"><code>sequence</code></a> method</a></h2><p>First, let&#39;s see how we do this with the <code>sequence</code> method. The first two parameters we give to the function will be the input genome we want to sequence, and the destination FASTQ file for output reads. Here we are setting:</p><ul><li>The number of genome copies in the molecule pool to 5000.</li><li>The average fragment size to 40000bp.</li><li>The sampling coverage to 30x.</li><li>The read length to <code>nothing</code>, which will make the sequencer read the whole length of any DNA fragment.</li><li>The per base read error rate to 0.1.</li><li>The fact we want paired-ends of fragments to be read (<code>paired</code>) to false.</li></ul><div><pre><code class="language-julia">sequence(&quot;ecoli-ref.fasta&quot;, &quot;longreads.fastq&quot;; ng = 5000, flen = 40000, cov = 30, rdlen = nothing, err = 0.1, paired = false)</code></pre><pre><code class="language-none">[ Info: - ✔ Created pool of 5000 copies of a 4639675bp genome
[ Info: - ✔ Created pool of fragments with an average length of 40000bp
[ Info: - ✔ Subsampled pool at 30X coverage (3479 molecules)
[ Info: - ✔ Created set of single-end reads
[ Info: - ✔ Applied sequencing errors at a per-base rate of 0.1
[ Info: - ✔ Wrote 3479 single end reads to FASTQ file</code></pre></div><h2><a class="nav-anchor" id="Using-the-Pseudoseq-API-1" href="#Using-the-Pseudoseq-API-1">Using the Pseudoseq API</a></h2><p>Here&#39;s how to achieve the same thing, using the Pseudoseq API. It is nessecery to use the API if you want to do something that is not anticipated by the available functionality of the <code>sequence</code> method: the cost of conveinience is fewer options.</p><p>Let&#39;s start with a pool of 5000 copies of a genome contained in a FASTA file:</p><div><pre><code class="language-julia">pool = makepool(&quot;ecoli-ref.fasta&quot;, 5000)</code></pre><pre><code class="language-none">Pool of 5000 molecules:
 All molecules are of the same size: 4639675</code></pre></div><p>Cut the pool of DNA into fragments of an average length of 40,000bp</p><div><pre><code class="language-julia">cutpool = fragment(pool, 40000)</code></pre><pre><code class="language-none">Pool of 575000 molecules:
 Maximum molecule size: 510928
 Average molecule size: 40345
 Minimum molecule size: 1</code></pre></div><p>Now we&#39;ll estimate the number of fragments we need to sample from the pool to achieve 30x coverage.</p><div><pre><code class="language-julia">genome_size = 4639675
expected_coverage = 30
readlength = 40000

N = needed_sample_size(expected_coverage, genome_size, readlength)

sampledpool = subsample(cutpool, N)</code></pre><pre><code class="language-none">Pool of 3479 molecules:
 Maximum molecule size: 387469
 Average molecule size: 40281
 Minimum molecule size: 27</code></pre></div><p>By using the <a href="../../../api/reads/#Pseudoseq.make_reads"><code>make_reads</code></a> function without specifying a read length, the function will generate reads from the entire length of each molecule in the pool. We do this to emulate what Nanopore sequencing is supposed to do: It takes an entire DNA fragment, feeds it through an electrically charged pore, producing a read for the entire fragment.</p><div><pre><code class="language-julia">se_reads = make_reads(SingleEnd, sampledpool)</code></pre></div><p>Long read sequencer have much higher error rates than short read sequencers so we use a error rate of 0.1.</p><div><pre><code class="language-julia">se_w_errs = mark_errors(se_reads, 0.1)</code></pre></div><p>Finally produce the ouput FASTQ file.</p><div><pre><code class="language-julia">generate(&quot;longreads.fastq&quot;, se_w_errs)#-</code></pre><pre><code class="language-none">[ Info: - ✔ Wrote 3479 single end reads to FASTQ file</code></pre></div><p><em>This page was generated using <a href="https://github.com/fredrikekre/Literate.jl">Literate.jl</a>.</em></p><footer><hr/><a class="previous" href="../pe-example/"><span class="direction">Previous</span><span class="title">Paired end reads</span></a><a class="next" href="../tg-example/"><span class="direction">Next</span><span class="title">Tagged paired end reads</span></a></footer></article></body></html>
