<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Tagged paired end reads · Pseudoseq.jl</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../../../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../../../assets/documenter.js"></script><script src="../../../../siteinfo.js"></script><script src="../../../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../../../"><img src="../../../../assets/logo.png" alt="Pseudoseq.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit">Pseudoseq.jl</span></div><form class="docs-search" action="../../../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../../../">Home</a></li><li><span class="tocitem">Manual</span><ul><li><a class="tocitem" href="../../../puzzlemaker/concepts/">PuzzleMaker</a></li><li><input class="collapse-toggle" id="menuitem-2-2" type="checkbox" checked/><label class="tocitem" for="menuitem-2-2"><span class="docs-label">Sequencing</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../concepts/">Core concepts &amp; workflow</a></li><li><input class="collapse-toggle" id="menuitem-2-2-2" type="checkbox" checked/><label class="tocitem" for="menuitem-2-2-2"><span class="docs-label">Examples</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../pe-example/">Paired end reads</a></li><li><a class="tocitem" href="../se-example/">Long single end reads</a></li><li class="is-active"><a class="tocitem" href>Tagged paired end reads</a><ul class="internal"><li><a class="tocitem" href="#Using-the-[sequence](@ref)-method-1"><span>Using the <code>sequence</code> method</span></a></li><li><a class="tocitem" href="#Using-the-Pseudoseq-API-1"><span>Using the Pseudoseq API</span></a></li><li><a class="tocitem" href="#Constructing-a-pipeline-of-Processors.-1"><span>Constructing a pipeline of <code>Processors</code>.</span></a></li></ul></li></ul></li></ul></li></ul></li><li><span class="tocitem">API</span><ul><li><input class="collapse-toggle" id="menuitem-3-1" type="checkbox"/><label class="tocitem" for="menuitem-3-1"><span class="docs-label">Sequencing</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../../../api/sequence/">sequence</a></li><li><a class="tocitem" href="../../../../api/pool/">Molecule Pool</a></li><li><a class="tocitem" href="../../../../api/reads/">Reads</a></li></ul></li><li><a class="tocitem" href="../../../../api/chromosome-blueprint/">Build-a-Genome</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Manual</a></li><li><a class="is-disabled">Sequencing</a></li><li><a class="is-disabled">Examples</a></li><li class="is-active"><a href>Tagged paired end reads</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Tagged paired end reads</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/master/examples/sequencing/tg-example.jl" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Example:-tagged-paired-end-reads-1"><a class="docs-heading-anchor" href="#Example:-tagged-paired-end-reads-1">Example: tagged paired-end reads</a><a class="docs-heading-anchor-permalink" href="#Example:-tagged-paired-end-reads-1" title="Permalink"></a></h1><p>This is an example generated from this source file: <a href="https://github.com/bioinfologics/Pseudoseq.jl/blob/masterexamples/sequencing/tg-example.jl"><code>tg-example.jl</code></a> You are seeing the online documentation version. The corresponding notebook can be found here: <a href="https://nbviewer.jupyter.org/github/bioinfologics/Pseudoseq.jl/blob/gh-pages/previews/PR15examples/sequencing/tg-example.ipynb"><code>tg-example.ipynb</code></a> and the script can be found here: <a href="../tg-example.jl"><code>tg-example.jl</code></a></p><p>Let&#39;s see how we might simulate something like an 10x sequencing experiment.</p><p>For this simulation script we will:</p><ol><li>Create a pool of 5000 copies of a reference genome.</li><li>Fragment the DNA molecules in the pool, to an average length of 40,000bp.</li><li>Tag the long molecules in the pool randomly with a set of 1,000,000 tags.</li><li>Fragment the molecules in the pool to an average length of 700bp.</li><li>Subsample the molecules in the pool to achieve approximatly 50x coverage.</li><li>Create a set of 250bp paired-end reads.</li><li>Apply errors to the paired-end reads at a rate of 0.001 (.1%).</li><li>Generate an output FASTQ file.</li></ol><pre><code class="language-julia">using Pseudoseq.Sequencing</code></pre><h2 id="Using-the-[sequence](@ref)-method-1"><a class="docs-heading-anchor" href="#Using-the-[sequence](@ref)-method-1">Using the <a href="man/sequencing/examples/@ref"><code>sequence</code></a> method</a><a class="docs-heading-anchor-permalink" href="#Using-the-[sequence](@ref)-method-1" title="Permalink"></a></h2><p>First, let&#39;s see how we do this with the <code>sequence</code> method. The first two parameters we give to the function will be the input genome we want to sequence, and the destination FASTQ file for output reads. Here we are setting:</p><ul><li>The number of genome copies in the molecule pool to 5,000.</li><li>The number of possible tags to one million.</li><li>The average fragment size, prior to tagging, to 40,000bp.</li><li>The average fragment size after tagging, to 700bp.</li><li>The sampling coverage to 50x.</li><li>The read length to 250bp.</li><li>The per base read error rate to 0.001.</li><li>The fact we want paired-ends of fragments to be read (<code>paired</code>) to true, which is the default.</li></ul><pre><code class="language-julia">sequence(&quot;ecoli-ref.fasta&quot;, &quot;tagged_reads.fastq&quot;; ng = 5000, tusize = 1000000, taggedflen = 40000, flen = 700, cov = 50, rdlen = 250, err = 0.1)</code></pre><pre><code class="language-none"></code></pre><h2 id="Using-the-Pseudoseq-API-1"><a class="docs-heading-anchor" href="#Using-the-Pseudoseq-API-1">Using the Pseudoseq API</a><a class="docs-heading-anchor-permalink" href="#Using-the-Pseudoseq-API-1" title="Permalink"></a></h2><p>Here&#39;s how to achieve the same thing, using the Pseudoseq API. It is nessecery to use the API if you want to do something that is not anticipated by the available functionality of the <code>sequence</code> method: the cost of conveinience is fewer options.</p><p>Let&#39;s start with a pool of 5000 copies of a genome contained in a FASTA file:</p><pre><code class="language-julia">dnapool = Molecules(&quot;ecoli-ref.fasta&quot;, 5000)</code></pre><pre><code class="language-none">Pool of 5000 DNA molecules:
 All molecules are of the same size: 4639675
</code></pre><p>Now let&#39;s cut up the molecules to an average length of 40,000bp</p><pre><code class="language-julia">cutpool = fragment(dnapool, 40000)</code></pre><pre><code class="language-none">Pool of 575000 DNA molecules:
 Maximum molecule size: 539514
 Average molecule size: 40345
 Minimum molecule size: 1
</code></pre><p>Ok, now we will tag these large fragments randomly. Once you tag a fragment in a universe, any other fragments that are derived from that tagged fragment will inherit the same tag.</p><pre><code class="language-julia">taggedpool = tag(cutpool, 1000000)</code></pre><pre><code class="language-none">Pool of 575000 DNA molecules:
 Maximum molecule size: 539514
 Average molecule size: 40345
 Minimum molecule size: 1
 Number of distinct tags: 437036
</code></pre><p>Here I&#39;m going to use a pool of 1,000,000 distinct tags. Which fragment gets a certain tag is random. The size of the tag pool, and the number of fragments in your universe will determine how likely it is that any two fragments get the same tag. Now we&#39;ll fragment the pool again</p><pre><code class="language-julia">taggedcutpool = fragment(taggedpool, 700)</code></pre><pre><code class="language-none">Pool of 32863285 DNA molecules:
 Maximum molecule size: 12463
 Average molecule size: 705
 Minimum molecule size: 1
 Number of distinct tags: 437036
</code></pre><p>Subsample the pool of tagged molecules.</p><pre><code class="language-julia">genome_size = 4639675
expected_coverage = 50
read_length = 250

N = needed_sample_size(expected_coverage, genome_size, read_length)
N = div(N, 2) # Divide by 2 as we&#39;re doing paired end sequencing.

sampledpool = subsample(taggedcutpool, N)</code></pre><pre><code class="language-none">Pool of 463967 DNA molecules:
 Maximum molecule size: 8987
 Average molecule size: 707
 Minimum molecule size: 1
 Number of distinct tags: 226750
</code></pre><p>Now let&#39;s make some 250bp tagged paired reads and generate some erroneous positions. We will construct a <code>FixedProbSubstitutions</code> function with a per base error probability of 0.001 and pass it to the <code>edit_substitutions</code> method.</p><pre><code class="language-julia">tagged_reads = paired_reads(sampledpool, 250)
f = FixedProbSubstitutions(0.001)
tagged_w_errs = edit_substitutions(f, tagged_reads)</code></pre><pre><code class="language-none"></code></pre><p>Output to FASTQ:</p><pre><code class="language-julia">generate(&quot;tagged_reads.fastq&quot;, tagged_w_errs)</code></pre><pre><code class="language-none">[ Info: - ✔ Wrote 655016 tagged paired end reads to FASTQ file</code></pre><h2 id="Constructing-a-pipeline-of-Processors.-1"><a class="docs-heading-anchor" href="#Constructing-a-pipeline-of-Processors.-1">Constructing a pipeline of <code>Processors</code>.</a><a class="docs-heading-anchor-permalink" href="#Constructing-a-pipeline-of-Processors.-1" title="Permalink"></a></h2><p>As a convenience, some users may prefer to use pipelines of <code>Processors</code> These behave like curried versions of the <code>Molecules</code> transformation methods. First let&#39;s define our starting <code>Molecules</code> pool:</p><pre><code class="language-julia">pool = Molecules(&quot;ecoli-ref.fasta&quot;, 5000)</code></pre><pre><code class="language-none">Pool of 5000 DNA molecules:
 All molecules are of the same size: 4639675
</code></pre><p>To make a Processor, use a <code>Molecules</code> transformation method, but do not provide a <code>Molecules</code> value as a first argument. So let&#39;s make Processors for each step of our paired-end, tagged-read, sequencing pipeline.</p><pre><code class="language-julia">cutter_a = fragment(40000)
tagger = tag(1000000)
cutter_b = fragment(700)
sampler = subsample(N) # Remember how to computed N previously.
mkreads = paired_reads(250)
adderr = make_substitutions(FixedProbSubstitutions(0.001))</code></pre><pre><code class="language-none">Pseudoseq.Sequencing.SubstitutionMaker{Pseudoseq.Sequencing.FixedProbSubstitutions}(Pseudoseq.Sequencing.FixedProbSubstitutions(0.001))</code></pre><p>Next we can construct the pipeline using standard julia function pipelining syntax:</p><pre><code class="language-julia">pool |&gt; cutter_a |&gt; tagger |&gt; cutter_b |&gt; sampler |&gt; mkreads |&gt; adderr |&gt; generate(&quot;tagged_reads.fastq&quot;)</code></pre><pre><code class="language-none">[ Info: Attaching 1000000 tags randomly to each molecule in pool...
[ Info: Done
[ Info: - ✔ Wrote 654420 tagged paired end reads to FASTQ file</code></pre><p>You can also compose the processors together into one whole function. Typing \circ in the julia repl and then hitting tab gives you the circular composition symbol. Note how pipelining above progresses from left to right, but composition is right to left in order.</p><pre><code class="language-julia">my_protocol = adderr ∘ mkreads ∘ sampler ∘ cutter_b ∘ tagger ∘ cutter_a

pool |&gt; my_protocol |&gt; generate(&quot;tagged_reads.fastq&quot;)</code></pre><pre><code class="language-none">[ Info: Attaching 1000000 tags randomly to each molecule in pool...
[ Info: Done
[ Info: - ✔ Wrote 653798 tagged paired end reads to FASTQ file</code></pre><hr/><p><em>This page was generated using <a href="https://github.com/fredrikekre/Literate.jl">Literate.jl</a>.</em></p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../se-example/">« Long single end reads</a><a class="docs-footer-nextpage" href="../../../../api/sequence/">sequence »</a></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Monday 18 May 2020 10:53">Monday 18 May 2020</span>. Using Julia version 1.2.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
