<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Paired end reads · Pseudoseq.jl</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../../../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../../../assets/documenter.js"></script><script src="../../../../siteinfo.js"></script><script src="../../../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../../../"><img src="../../../../assets/logo.png" alt="Pseudoseq.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit">Pseudoseq.jl</span></div><form class="docs-search" action="../../../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../../../">Home</a></li><li><span class="tocitem">Manual</span><ul><li><a class="tocitem" href="../../../puzzlemaker/concepts/">PuzzleMaker</a></li><li><input class="collapse-toggle" id="menuitem-2-2" type="checkbox" checked/><label class="tocitem" for="menuitem-2-2"><span class="docs-label">Sequencing</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../concepts/">Core concepts &amp; workflow</a></li><li><input class="collapse-toggle" id="menuitem-2-2-2" type="checkbox" checked/><label class="tocitem" for="menuitem-2-2-2"><span class="docs-label">Examples</span><i class="docs-chevron"></i></label><ul class="collapsed"><li class="is-active"><a class="tocitem" href>Paired end reads</a><ul class="internal"><li><a class="tocitem" href="#Using-the-[sequence](@ref)-method-1"><span>Using the <code>sequence</code> method</span></a></li><li><a class="tocitem" href="#Using-the-Molecules-tranformation-methods.-1"><span>Using the <code>Molecules</code> tranformation methods.</span></a></li><li><a class="tocitem" href="#Constructing-a-pipeline-of-Processors.-1"><span>Constructing a pipeline of <code>Processors</code>.</span></a></li></ul></li><li><a class="tocitem" href="../se-example/">Long single end reads</a></li><li><a class="tocitem" href="../tg-example/">Tagged paired end reads</a></li></ul></li></ul></li></ul></li><li><span class="tocitem">API</span><ul><li><input class="collapse-toggle" id="menuitem-3-1" type="checkbox"/><label class="tocitem" for="menuitem-3-1"><span class="docs-label">Sequencing</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../../../api/sequence/">sequence</a></li><li><a class="tocitem" href="../../../../api/pool/">Molecule Pool</a></li><li><a class="tocitem" href="../../../../api/reads/">Reads</a></li></ul></li><li><a class="tocitem" href="../../../../api/chromosome-blueprint/">Build-a-Genome</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Manual</a></li><li><a class="is-disabled">Sequencing</a></li><li><a class="is-disabled">Examples</a></li><li class="is-active"><a href>Paired end reads</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Paired end reads</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/master/examples/sequencing/pe-example.jl" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Example:-paired-end-sequencing-1"><a class="docs-heading-anchor" href="#Example:-paired-end-sequencing-1">Example: paired-end sequencing</a><a class="docs-heading-anchor-permalink" href="#Example:-paired-end-sequencing-1" title="Permalink"></a></h1><p>This is an example generated from this source file: <a href="https://github.com/bioinfologics/Pseudoseq.jl/blob/masterexamples/sequencing/pe-example.jl"><code>pe-example.jl</code></a> You are seeing the online documentation version. The corresponding notebook can be found here: <a href="https://nbviewer.jupyter.org/github/bioinfologics/Pseudoseq.jl/blob/gh-pages/previews/PR15examples/sequencing/pe-example.ipynb"><code>pe-example.ipynb</code></a> and the script can be found here: <a href="../pe-example.jl"><code>pe-example.jl</code></a></p><p>For the simulation we are going to:</p><ol><li>Create a pool of 5000 copies of a reference genome.</li><li>Fragment the DNA molecules in the pool, to an average length of 700bp.</li><li>Subsample the molecules in the pool to achieve approximatly 50x coverage.</li><li>Create a set of 250bp paired-end reads.</li><li>Apply errors to the paired-end reads at a rate of 0.001 (.1%).</li><li>Generate an output FASTQ file.</li></ol><pre><code class="language-julia">using Pseudoseq.Sequencing</code></pre><h2 id="Using-the-[sequence](@ref)-method-1"><a class="docs-heading-anchor" href="#Using-the-[sequence](@ref)-method-1">Using the <a href="man/sequencing/examples/@ref"><code>sequence</code></a> method</a><a class="docs-heading-anchor-permalink" href="#Using-the-[sequence](@ref)-method-1" title="Permalink"></a></h2><p>First, let&#39;s see how we do this with the <code>sequence</code> method. The first two parameters we give to the function will be the input genome we want to sequence, and the destination FASTQ file for output reads. Here we are setting:</p><ul><li>The number of genome copies in the molecule pool to 5000.</li><li>The average fragment size to 700bp.</li><li>The sampling coverage to 50x.</li><li>The read length to 250bp.</li><li>The per base read error rate to 0.001.</li><li>The fact we want paired-ends of fragments to be read (<code>paired</code>) to true.</li></ul><pre><code class="language-julia">sequence(&quot;ecoli-ref.fasta&quot;, &quot;pe-reads.fastq&quot;; ng = 5000, flen = 700, cov = 50, paired = true, rdlen = 250, err = 0.001)</code></pre><pre><code class="language-none"></code></pre><h2 id="Using-the-Molecules-tranformation-methods.-1"><a class="docs-heading-anchor" href="#Using-the-Molecules-tranformation-methods.-1">Using the <code>Molecules</code> tranformation methods.</a><a class="docs-heading-anchor-permalink" href="#Using-the-Molecules-tranformation-methods.-1" title="Permalink"></a></h2><p>Here&#39;s how to achieve the same thing, using the Pseudoseq API. It is nessecery to use the API if you want to do something that is not anticipated by the available functionality of the <code>sequence</code> method: the cost of conveinience is fewer options.</p><p>Starting with a FASTA formatted file containing the genome we want to sequence, we create a pool with 5000 copies of the genome.</p><pre><code class="language-julia">pool = Molecules(&quot;ecoli-ref.fasta&quot;, 5000)</code></pre><pre><code class="language-none">Pool of 5000 DNA molecules:
 All molecules are of the same size: 4639675
</code></pre><p>Next we use the fragment function to make a pool of shorter DNA molecules.</p><pre><code class="language-julia">cutpool = fragment(pool, 700)</code></pre><pre><code class="language-none">Pool of 33139991 DNA molecules:
 Maximum molecule size: 13224
 Average molecule size: 700
 Minimum molecule size: 1
</code></pre><p>We need to determine the number of molecules to sample, and subsample the pool:</p><pre><code class="language-julia">genome_size = 4639675
expected_coverage = 50
read_length = 250

N = needed_sample_size(expected_coverage, genome_size, read_length)
N = div(N, 2) # Divide by 2 as we&#39;re doing paired end sequencing.

sampledpool = subsample(cutpool, N)</code></pre><pre><code class="language-none">Pool of 463967 DNA molecules:
 Maximum molecule size: 10721
 Average molecule size: 699
 Minimum molecule size: 1
</code></pre><p>We now want to create a set of paired-end reads. We want our reads to be 250bp in length.</p><pre><code class="language-julia">pe_reads = paired_reads(sampledpool, 250)</code></pre><pre><code class="language-none"></code></pre><p>Now we have some reads, we should mark positions in the reads that are destined to be errors in the output FASTQ. We will construct a <code>FixedProbSubstitutions</code> function with a per base error probability of 0.001 and pass it to the <code>edit_substitutions</code> method.</p><pre><code class="language-julia">f = FixedProbSubstitutions(0.001)
pe_w_errs = edit_substitutions(f, pe_reads)</code></pre><pre><code class="language-none"></code></pre><p>Now we have some paired end reads and have marked some positions as errors, we can generate FASTQ files.</p><pre><code class="language-julia">generate(&quot;pe-reads.fastq&quot;, pe_w_errs)</code></pre><pre><code class="language-none">[ Info: - ✔ Wrote 649398 paired end reads to FASTQ file</code></pre><h2 id="Constructing-a-pipeline-of-Processors.-1"><a class="docs-heading-anchor" href="#Constructing-a-pipeline-of-Processors.-1">Constructing a pipeline of <code>Processors</code>.</a><a class="docs-heading-anchor-permalink" href="#Constructing-a-pipeline-of-Processors.-1" title="Permalink"></a></h2><p>As a convenience, some users may prefer to use pipelines of <code>Processors</code> These behave like curried versions of the <code>Molecules</code> transformation methods. First let&#39;s define our starting <code>Molecules</code> pool:</p><pre><code class="language-julia">pool = Molecules(&quot;ecoli-ref.fasta&quot;, 5000)</code></pre><pre><code class="language-none">Pool of 5000 DNA molecules:
 All molecules are of the same size: 4639675
</code></pre><p>To make a Processor, use a <code>Molecules</code> transformation method, but do not provide a <code>Molecules</code> value as a first argument. So let&#39;s make Processors for each step of our paired end sequencing pipeline.</p><pre><code class="language-julia">cutter = fragment(700)
sampler = subsample(N) # Remember how to computed N previously.
mkreads = paired_reads(250)
adderr = make_substitutions(FixedProbSubstitutions(0.001))</code></pre><pre><code class="language-none">Pseudoseq.Sequencing.SubstitutionMaker{Pseudoseq.Sequencing.FixedProbSubstitutions}(Pseudoseq.Sequencing.FixedProbSubstitutions(0.001))</code></pre><p>Next we can construct the pipeline using standard julia function pipelining syntax:</p><pre><code class="language-julia">pool |&gt; cutter |&gt; sampler |&gt; mkreads |&gt; adderr |&gt; generate(&quot;pe-reads.fastq&quot;)</code></pre><pre><code class="language-none">[ Info: - ✔ Wrote 648614 paired end reads to FASTQ file</code></pre><p>You can also compose the processors together into one whole function. Typing \circ in the julia repl and then hitting tab gives you the circular composition symbol. Note how pipelining above progresses from left to right, but composition is right to left in order.</p><pre><code class="language-julia">my_protocol = adderr ∘ mkreads ∘ sampler ∘ cutter

pool |&gt; my_protocol |&gt; generate(&quot;pe-reads.fastq&quot;)</code></pre><pre><code class="language-none">[ Info: - ✔ Wrote 649192 paired end reads to FASTQ file</code></pre><hr/><p><em>This page was generated using <a href="https://github.com/fredrikekre/Literate.jl">Literate.jl</a>.</em></p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../../concepts/">« Core concepts &amp; workflow</a><a class="docs-footer-nextpage" href="../se-example/">Long single end reads »</a></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Monday 18 May 2020 01:25">Monday 18 May 2020</span>. Using Julia version 1.2.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
