<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Chromosome Blueprint · Pseudoseq.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link href="../../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><a href="../../index.html"><img class="logo" src="../../assets/logo.png" alt="Pseudoseq.jl logo"/></a><h1>Pseudoseq.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../../search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../../">Home</a></li><li><span class="toctext">Build-a-Genome</span><ul><li><a class="toctext" href="../../build-a-genome/">Core concepts &amp; workflow</a></li><li><span class="toctext">Examples</span><ul><li><a class="toctext" href="../../examples/build-a-yeast/">Build-a-Yeast</a></li></ul></li></ul></li><li><span class="toctext">Sequencing</span><ul><li><a class="toctext" href="../../sequencing/">Core concepts &amp; workflow</a></li><li><span class="toctext">Examples</span><ul><li><a class="toctext" href="../../examples/pe-example/">Paired end reads</a></li><li><a class="toctext" href="../../examples/se-example/">Long single end reads</a></li><li><a class="toctext" href="../../examples/tg-example/">Tagged paired end reads</a></li></ul></li></ul></li><li><span class="toctext">API</span><ul><li><a class="toctext" href="../sequence/">Sequencing</a></li><li><a class="toctext" href="../pool/">Molecule Pool</a></li><li><a class="toctext" href="../reads/">Reads</a></li><li class="current"><a class="toctext" href>Chromosome Blueprint</a><ul class="internal"><li><a class="toctext" href="#Exported-functions-1">Exported functions</a></li></ul></li></ul></li></ul></nav><article id="docs"><header><nav><ul><li>API</li><li><a href>Chromosome Blueprint</a></li></ul><a class="edit-page" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/master/docs/src/api/chromosome-blueprint.md"><span class="fa"></span> Edit on GitHub</a></nav><hr/><div id="topbar"><span>Chromosome Blueprint</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="API:-Build-a-Genome-1" href="#API:-Build-a-Genome-1">API: Build-a-Genome</a></h1><h2><a class="nav-anchor" id="Exported-functions-1" href="#Exported-functions-1">Exported functions</a></h2><h3><a class="nav-anchor" id="Making-an-empty-chromosome-blueprint-1" href="#Making-an-empty-chromosome-blueprint-1">Making an empty chromosome blueprint</a></h3><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Pseudoseq.plan_chrom" href="#Pseudoseq.plan_chrom"><code>Pseudoseq.plan_chrom</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Create an empty blueprint for <code>n</code> copies of a chromosome of <code>len</code> base pairs.</p></div></div><a class="source-link" target="_blank" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/d5ba2db025a95293f6851b31a6f7c110d3c268d2/src/build_a_genome/plan/chromosomes.jl#L1">source</a></section><h3><a class="nav-anchor" id="Planning-motif-repetitions-along-chromosomes-1" href="#Planning-motif-repetitions-along-chromosomes-1">Planning motif repetitions along chromosomes</a></h3><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Pseudoseq.plan_repetition" href="#Pseudoseq.plan_repetition"><code>Pseudoseq.plan_repetition</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">plan_repetition(cb::ChromosomeBlueprint, from::UnitRange{Int}, to::UnitRange{Int})</code></pre><p>Plan a repetition in a chromosome, where the bases in the <code>from</code> region of the chromosome, are guaranteed to occur again in the <code>to</code> region of the chromosome.</p><p>Every copy of the chromosome will have this same repetition.</p><p>Creates a new chromosome blueprint, based on the input blueprint <code>cb</code>.</p><div class="admonition note"><div class="admonition-title">Note</div><div class="admonition-text"><p>In the new blueprint, the <code>to</code> region of the planned chromosome will be consumed, and cannot be used to plan any other subsequently added features.</p></div></div></div></div><a class="source-link" target="_blank" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/d5ba2db025a95293f6851b31a6f7c110d3c268d2/src/build_a_genome/plan/repetitions.jl#L4-L18">source</a><div><div><pre><code class="language-none">plan_repetition(cb::ChromosomeBlueprint, from::Int, to::Int, size::Int)</code></pre><p>Plan a repetition in a chromosome, where the bases in the <code>from</code>:(<code>from</code> + <code>size</code> - 1) region of the chromosome, are guaranteed to occur again in the <code>to</code>:(<code>to</code> + <code>size</code> - 1) region of the chromosome.</p><p>Every copy of the chromosome will have this same repetition.</p><p>Creates a new chromosome blueprint, based on the input blueprint <code>cb</code>.</p></div></div><a class="source-link" target="_blank" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/d5ba2db025a95293f6851b31a6f7c110d3c268d2/src/build_a_genome/plan/repetitions.jl#L30-L40">source</a><div><div><pre><code class="language-none">plan_repetition(cb::ChromosomeBlueprint, intervals::Vector{UnitRange{Int}})</code></pre><p>A conveinience method of <code>plan_repetition</code>. Designed to ease the process of planing a series of repetitions in a chromosome.</p><p>Every copy of the chromosome will have these same repetitions.</p><p>Creates a new chromosome blueprint, based on the input blueprint <code>cb</code>.</p><div class="admonition tip"><div class="admonition-title">Tip</div><div class="admonition-text"><p>Use the <a href="#Pseudoseq.suggest_regions"><code>suggest_regions</code></a> function to help decide on a set of sites to make heterozygous.</p></div></div><div class="admonition note"><div class="admonition-title">Note</div><div class="admonition-text"><p>The number of intervals provided must be an even number. This is because intervals 1 &amp; 2  define the first repeat, intervals 3 &amp; 4 define the second, and so on.</p></div></div><div class="admonition note"><div class="admonition-title">Note</div><div class="admonition-text"><p>In the new blueprint, the regions the repetitions occupy, have been consumed, and cannot be used to plan any other subsequently added features.</p></div></div></div></div><a class="source-link" target="_blank" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/d5ba2db025a95293f6851b31a6f7c110d3c268d2/src/build_a_genome/plan/repetitions.jl#L48-L70">source</a></section><h3><a class="nav-anchor" id="Planning-heterozygosity-between-chromosome-copies-1" href="#Planning-heterozygosity-between-chromosome-copies-1">Planning heterozygosity between chromosome copies</a></h3><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Pseudoseq.plan_het" href="#Pseudoseq.plan_het"><code>Pseudoseq.plan_het</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">plan_het(cb::ChromosomeBlueprint, pos::Int, alleles::Vector{DNA})</code></pre><p>Plan heterozygosity between copies of a chromosome at position <code>pos</code>.</p><p>The <code>alleles</code> vector must contain a nucleotide for each copy of the chromosome.</p><p>For example if you provided a vector of <code>[DNA_A, DNA_C]</code>, for the heterozygous site you&#39;re defining at <code>pos</code>, the first copy of the chromosome will have an <code>A</code>, and the second copy of the chromosome will have a <code>C</code>.</p><p>Creates a new chromosome blueprint, based on the input blueprint <code>cb</code>.</p><div class="admonition tip"><div class="admonition-title">Tip</div><div class="admonition-text"><p>Use the <a href="#Pseudoseq.suggest_alleles"><code>suggest_alleles</code></a> function to help decide on a set of alleles to use.</p></div></div><div class="admonition note"><div class="admonition-title">Note</div><div class="admonition-text"><p>In the new blueprint, the positions that were used to plan the heterozygosity, have been consumed, and cannot be used to plan any other subsequently added features.</p></div></div></div></div><a class="source-link" target="_blank" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/d5ba2db025a95293f6851b31a6f7c110d3c268d2/src/build_a_genome/plan/polymorphism.jl#L6-L28">source</a><div><div><pre><code class="language-none">plan_het(cb::ChromosomeBlueprint, pos, alle...)</code></pre><p>A generic method of <code>plan_het</code>.</p><p>Creates a new chromosome blueprint, based on the input blueprint <code>cb</code>.</p><p>The <code>pos</code> argument determines which sites in a chromosome are made heterozygous, and depending on the type of argument provided, the behaviour differs slightly:</p><ul><li><p>If <code>pos</code> is an integer, then that many available sites are selected at random to be heterozygous.</p></li><li><p>If <code>pos</code> is a float, then it is treated as a proportion of the length of the, chromosome, which is converted into an integer, and that many sites are selected at random to be heterozygous.</p></li><li><p>If <code>pos</code> is a vector of integers, it is trated as a list of positions the user wants to be heterozygous.</p></li></ul><p>The <code>alle...</code> argument is a vararg argument which defines the how bases will be  allocated at each heterozygous site.</p><p>Depending on the type of the argument provided, the behaviour differs slightly, based on the different methods of <a href="#Pseudoseq.suggest_alleles"><code>suggest_alleles</code></a>:</p><ul><li><p>If <code>alle...</code> is a single integer value. It is taken to mean the number of states each heterozygous site has. At a minumum this number must be two, as a site with only one state is not heterozygous by definition. For each site, which chromosome copies get which state is determined at random.</p></li><li><p>If <code>alle...</code> is a vector of integers (one integer for every chromosome copy), then the nth value in the vector dictates the group the nth chromosome copy belongs to. For example for a triploid, the vector <code>[2, 1, 2]</code> means that there are two groups (group 1 and group 2), and that the second chromosome copy of the three belongs to group 1, and the other two copies belong to group</p><ol><li>Chromosome copies that belong to the same group will get the same base at</li></ol><p>each heterozygous site. In our example, copies 1 and 3 will get the same base, as they have been given the same group number. Chromosome copy 2 will end up with a different base. Which base corresponds to which group number will be determined at random.</p></li><li><p>If <code>alle...</code> is multiple vectors of integers, then each vector constitutes a group, and each value in the vector determines which copies are in that group. It is alternative way to giving the same information as you can with a single vector (as above). E.g. consider the example of <code>[2, 1, 2]</code> from the previous point. This can be expressed as two vectors of <code>[2] and [1, 3]</code>: The first vector contains a number indicating the second chromosome copy, and the second vector contains numbers representing the first and third chromosome copy.</p></li><li><p>If <code>alle...</code> is a vector of nucleotide vectors, then the nth vector of nucleotides determines which chromosome copy recieves which base at the nth heterozygous position. For example, for a triploid, <code>[[DNA_A, DNA_A, DNA_T], [DNA_G, DNA_C, DNA_C]]</code> would mean that at the first heterozygous site, the first and second copies of the chromosome would recieve an <code>DNA_A</code> base, and the third copy would recieve a <code>DNA_T</code> base. For the second heterozygous position, the first copy of the chromosome would recieve a DNA<em>G base, and the other two copies will recieve a `DNA</em>C` base.</p></li></ul><p>(See also: <a href="#Pseudoseq.plan_repetition"><code>plan_repetition</code></a>)</p></div></div><a class="source-link" target="_blank" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/d5ba2db025a95293f6851b31a6f7c110d3c268d2/src/build_a_genome/plan/polymorphism.jl#L159-L220">source</a></section><h3><a class="nav-anchor" id="Fabricate-1" href="#Fabricate-1">Fabricate</a></h3><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Pseudoseq.fabricate" href="#Pseudoseq.fabricate"><code>Pseudoseq.fabricate</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">fabricate(file::String, cb...)</code></pre><p>Fabricate the sequence(s) planned in a number chromosome blueprints.</p><p>The fabricated sequences will be written out to the FASTA formatted file <code>file</code>.</p><p><code>cb...</code> should be provided as a series of blueprint, seed-sequence pairs. E.g. <code>fabricate(&quot;mygenome.fasta&quot;, chr1plan =&gt; ch1seq, ch2plan =&gt; ch2seq)</code>.</p></div></div><a class="source-link" target="_blank" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/d5ba2db025a95293f6851b31a6f7c110d3c268d2/src/build_a_genome/fabrication.jl#L1-L10">source</a><div><div><pre><code class="language-none">fabricate(fw::FASTA.Writer, cb...)</code></pre><p>Fabricate the sequence(s) planned in a number chromosome blueprints.</p><p>The fabricated sequences will be written out to the FASTA formatted file <code>fw</code>.</p><p><code>cb...</code> should be provided as a series of blueprint, seed-sequence pairs. E.g. <code>fabricate(&quot;mygenome.fasta&quot;, chr1plan =&gt; ch1seq, ch2plan =&gt; ch2seq)</code>.</p></div></div><a class="source-link" target="_blank" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/d5ba2db025a95293f6851b31a6f7c110d3c268d2/src/build_a_genome/fabrication.jl#L18-L27">source</a><div><div><pre><code class="language-none">fabricate(cb::ChromosomeBlueprint)</code></pre><p>Fabricate the sequence(s) planned in the chromosome blueprint.</p><p>A random DNA sequence will be generated to use as a seed sequence.</p><p>A sequence will be built for each chromosome copy in the blueprint.</p></div></div><a class="source-link" target="_blank" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/d5ba2db025a95293f6851b31a6f7c110d3c268d2/src/build_a_genome/fabrication.jl#L41-L49">source</a><div><div><pre><code class="language-none">fabricate(cb::ChromosomeBlueprint, seed::BioSequence{DNAAlphabet{2}})</code></pre><p>Fabricate a DNA sequence by applying the planned features in a chromosome blueprint, to some initial starting seed sequence.</p><p>A sequence will be built for each chromosome copy in the blueprint.</p></div></div><a class="source-link" target="_blank" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/d5ba2db025a95293f6851b31a6f7c110d3c268d2/src/build_a_genome/fabrication.jl#L55-L62">source</a></section><h3><a class="nav-anchor" id="Utility-functions-1" href="#Utility-functions-1">Utility functions</a></h3><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Pseudoseq.suggest_regions" href="#Pseudoseq.suggest_regions"><code>Pseudoseq.suggest_regions</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">suggest_regions(cp::ChromosomeBlueprint, size::Int, n::Int)</code></pre><p>A useful utility function to assist planning features in a chromosome blueprint.</p><p>This method returns a vector of non-overlapping, regions of the chromosome planned represented by the ChromosomeBlueprint <code>cb</code>.</p><p>These regions are free regions: they are untouched by any other planned features, and so may be used when planning other features (See also: <a href="#Pseudoseq.plan_repetition"><code>plan_repetition</code></a>, <a href="#Pseudoseq.plan_het"><code>plan_het</code></a>).</p><p>The regions will be <code>size</code>bp in length.</p><div class="admonition warn"><div class="admonition-title">Warn</div><div class="admonition-text"><p>This function was designed for use interactively in a julia session.</p><p>If this method cannot find <code>n</code> free regions of the size you&#39;ve asked for, it will still give an output vector containing the regions it did manage to find, but it will issue a warning to the terminal. This will get increasingly likely as you fill the chromosome blueprint up with features.</p><p>If you are using this method in a script or program where you depend on a reliable number of regions, either add a check to make sure you got the number of regions you need, or use this method interactively, and hard code an appropriate output into your script.</p></div></div></div></div><a class="source-link" target="_blank" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/d5ba2db025a95293f6851b31a6f7c110d3c268d2/src/build_a_genome/chrom_blueprint.jl#L205-L230">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Pseudoseq.suggest_alleles" href="#Pseudoseq.suggest_alleles"><code>Pseudoseq.suggest_alleles</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">suggest_alleles(groups::Vector{Int})</code></pre><p>Suggest an allele pattern for a single heterozygous site.</p><p>The <code>groups</code> vector should have a number of  The nth value in the <code>groups</code> designates a group to the nth chromosome copy.</p><p>E.g. consider for a triploid, the vector <code>[2, 1, 2]</code> means that there are two groups (group 1 and group 2), and that the second chromosome copy of the three belongs to group 1, and the other two copies belong to group 2.</p><p>Chromosome copies that belong to the same group will get the same base at each heterozygous site.</p><p>For the vector <code>[2, 1, 2]</code>, this method might return a suggested allele pattern such as <code>[DNA_A, DNA_G, DNA_A]</code>.</p><div class="admonition note"><div class="admonition-title">Note</div><div class="admonition-text"><p>Which base corresponds to which group number will be determined at random.</p></div></div></div></div><a class="source-link" target="_blank" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/d5ba2db025a95293f6851b31a6f7c110d3c268d2/src/build_a_genome/plan/polymorphism.jl#L49-L69">source</a><div><div><pre><code class="language-none">suggest_alleles(groups::Vector{Int}...)</code></pre><p>Suggest an allele pattern for a single heterozygous site.</p><p>Multiple integer vectors make up <code>groups</code>.</p><p>Each vector defines a group, and each value in a vector determines which copies are allocated to that group.</p><p>E.g. consider for a triploid, two vectors of <code>[2] and [1, 3]</code> passed as arguments to <code>groups</code>. They define two groups, the first vector contains a number denoting the second chromosome copy, and he second vector contains numbers representing the first and third chromosome copy.</p><p>For the vectors <code>[2]</code> and <code>[1, 3]</code>, this method might return a suggested allele pattern such as <code>[DNA_A, DNA_G, DNA_A]</code>.</p><div class="admonition note"><div class="admonition-title">Note</div><div class="admonition-text"><p>Which base corresponds to which group number will be determined at random.</p></div></div></div></div><a class="source-link" target="_blank" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/d5ba2db025a95293f6851b31a6f7c110d3c268d2/src/build_a_genome/plan/polymorphism.jl#L76-L96">source</a><div><div><pre><code class="language-none">suggest_alleles(ncopies::Int, ngroups::Int)</code></pre><p>Suggest an allele pattern for a single heterozygous site.</p><p>By providing a number representing the number of copies of a chromosome, and a number representing the number of groups (which at a minimum must be 2). This method of <code>suggest_alleles</code> will randomly allocate the <code>ncopies</code> of the chromosome into the <code>ngroups</code> groups.</p><p>For example, if you ask for a suggested allele pattern for 3 copies of a chromosome, and 2 groups, this method might return a suggested pattern of <code>[DNA_G, DNA_A, DNA_G]</code>.</p><div class="admonition note"><div class="admonition-title">Note</div><div class="admonition-text"><p>Which base corresponds to which group number will be determined at random.</p></div></div></div></div><a class="source-link" target="_blank" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/d5ba2db025a95293f6851b31a6f7c110d3c268d2/src/build_a_genome/plan/polymorphism.jl#L108-L123">source</a><div><div><pre><code class="language-none">suggest_alleles(npositions::Int, args...)</code></pre><p>Suggest allele patterns for numerous heterozygous sites.</p><p>This method of <code>suggest_alleles</code> calls another <code>suggest_alleles</code> with <code>args...</code> <code>npositions</code> times, and returns a vector of the results of each call.</p></div></div><a class="source-link" target="_blank" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/d5ba2db025a95293f6851b31a6f7c110d3c268d2/src/build_a_genome/plan/polymorphism.jl#L131-L138">source</a></section><footer><hr/><a class="previous" href="../reads/"><span class="direction">Previous</span><span class="title">Reads</span></a></footer></article></body></html>
