<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Build-a-Yeast · Pseudoseq.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link href="../../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><a href="../../index.html"><img class="logo" src="../../assets/logo.png" alt="Pseudoseq.jl logo"/></a><h1>Pseudoseq.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../../search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../../">Home</a></li><li><span class="toctext">Build-a-Genome</span><ul><li><a class="toctext" href="../../build-a-genome/">Core concepts &amp; workflow</a></li><li><span class="toctext">Examples</span><ul><li class="current"><a class="toctext" href>Build-a-Yeast</a><ul class="internal"><li><a class="toctext" href="#Load-the-seed-sequence-1">Load the seed sequence</a></li><li><a class="toctext" href="#Designing-a-diploid-1">Designing a diploid</a></li><li><a class="toctext" href="#Designing-a-triploid-1">Designing a triploid</a></li><li><a class="toctext" href="#Designing-a-tetraploid-genome-1">Designing a tetraploid genome</a></li></ul></li></ul></li></ul></li><li><span class="toctext">Sequencing</span><ul><li><a class="toctext" href="../../sequencing/">Core concepts &amp; workflow</a></li><li><span class="toctext">Examples</span><ul><li><a class="toctext" href="../pe-example/">Paired end reads</a></li><li><a class="toctext" href="../se-example/">Long single end reads</a></li><li><a class="toctext" href="../tg-example/">Tagged paired end reads</a></li></ul></li></ul></li><li><span class="toctext">API</span><ul><li><a class="toctext" href="../../api/sequence/">Sequencing</a></li><li><a class="toctext" href="../../api/pool/">Molecule Pool</a></li><li><a class="toctext" href="../../api/reads/">Reads</a></li><li><a class="toctext" href="../../api/chromosome-blueprint/">Chromosome Blueprint</a></li></ul></li></ul></nav><article id="docs"><header><nav><ul><li>Build-a-Genome</li><li>Examples</li><li><a href>Build-a-Yeast</a></li></ul><a class="edit-page" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/master/examples/build-a-yeast.jl"><span class="fa"></span> Edit on GitHub</a></nav><hr/><div id="topbar"><span>Build-a-Yeast</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="Example:-Build-a-Yeast-1" href="#Example:-Build-a-Yeast-1">Example: Build-a-Yeast</a></h1><p>This is an example generated from this source file: <a href="https://github.com/bioinfologics/Pseudoseq.jl/blob/master/examples/build-a-yeast.jl"><code>build-a-yeast.jl</code></a> You are seeing the online documentation version. The corresponding notebook can be found here: <a href="https://nbviewer.jupyter.org/github/bioinfologics/Pseudoseq.jl/blob/gh-pages/dev/examples/build-a-yeast.ipynb"><code>build-a-yeast.ipynb</code></a> and the script can be found here: <a href="../build-a-yeast.jl"><code>build-a-yeast.jl</code></a></p><p>We are going to use Pseudoseq to create a simple set of fake genomes, based on chromosome 1 of the reference genome for yeast.</p><ol><li>Load the sequence of chromosome 1 of the yeast reference from FASTA file.</li><li>Create blank chromosome blueprints for different genome designs.</li><li>Edit the blueprints to achive a: a. Diploid chromosome b. Triploid chromosome c. Tetraploid chromosome</li></ol><div><pre><code class="language-julia">using BioSequences, Pseudoseq</code></pre></div><h2><a class="nav-anchor" id="Load-the-seed-sequence-1" href="#Load-the-seed-sequence-1">Load the seed sequence</a></h2><p>First load the sequence from FASTA file. This uses tools from BioSequences, which is a dependency of Pseudoseq and so the user should have it available, as julia&#39;s package manager would have installed any dependencies. The lines below open a fasta file, with a fasta reader, and load a single FASTA record, and get its sequence.</p><div><pre><code class="language-julia">refseq = open(FASTA.Reader, &quot;yeast-chr1.fasta&quot;) do rdr
    FASTA.sequence(BioSequence{DNAAlphabet{2}}, read(rdr))
end

reflen = length(refseq)</code></pre><pre><code class="language-none">230218</code></pre></div><p>This results in a 230,218nt long sequence called <code>refseq</code>.</p><h2><a class="nav-anchor" id="Designing-a-diploid-1" href="#Designing-a-diploid-1">Designing a diploid</a></h2><p>First lets make a plan for a diploid genome with 1% heterozygosity. Pseudoseq has a lot of flexibilty in how you can define the heterozygosity, but for ease here I&#39;m going to let it decide where the sites should be, and what the alleles at each site should be, by calling <a href="../../api/chromosome-blueprint/#Pseudoseq.plan_het"><code>plan_het</code></a>, providing just the proportion of sites I want to be heterozygous, and the number of alleles I want at each of these heterozygous sites (2).</p><div><pre><code class="language-julia">dploid = plan_chrom(reflen, 2)
dploidhet = plan_het(dploid, .01, 2)</code></pre><pre><code class="language-none">Specification for a chromosome:
 Number of copies: 2
 Length of chromosomes: 230218
 227916bp available in the following stretches:</code></pre></div><p>Now fabricate the FASTA, with <a href="../../api/chromosome-blueprint/#Pseudoseq.fabricate"><code>fabricate</code></a>, providing a filename, and then a series of <code>blueprint =&gt; seed</code> pairs. In this case we just give 1 blueprint-seed</p><div><pre><code class="language-julia"># pair.

fabricate(&quot;build-a-yeast-di.fasta&quot;, dploidhet =&gt; refseq)</code></pre><pre><code class="language-none">[ Info: Opening FASTA file at: build-a-yeast-di.fasta
[ Info: Fabricated 1 chromosomes and written to file</code></pre></div><p>Thats it for the diploid genome.</p><h2><a class="nav-anchor" id="Designing-a-triploid-1" href="#Designing-a-triploid-1">Designing a triploid</a></h2><p>Now let&#39;s make a triploid genome. This should also be simple enough. It&#39;s much like the process for making the diploid genome, but specifying 3 chromosome copies instead of 2.</p><div><pre><code class="language-julia">trploid = plan_chrom(reflen, 3)</code></pre><pre><code class="language-none">Specification for a chromosome:
 Number of copies: 3
 Length of chromosomes: 230218
 230218bp available in the following stretches:</code></pre></div><p>When I add some heterozygous sites, just like with the diploid case, I&#39;m going to simply ask for a proportion of heterozygous sites (and let Pseudoseq decide where they should occur. Like last time also I&#39;m going to make all the sites have 2 alleles, so for every heterozygous site, two of the three chromosome copies willl have the same base, and one of the copies will be different. There are other ways to define some heterozygous sites for example all 3 chromosome copies could have different bases</p><div><pre><code class="language-julia">trploidhet = plan_het(trploid, .01, 2)</code></pre><pre><code class="language-none">Specification for a chromosome:
 Number of copies: 3
 Length of chromosomes: 230218
 227916bp available in the following stretches:</code></pre></div><p>Now fabricate the FASTA, with <code>fabricate</code>, providing a filename, and then a series of <code>blueprint =&gt; seed</code> pairs. In this case we just give 1 blueprint-seed</p><div><pre><code class="language-julia"># pair.

fabricate(&quot;build-a-yeast-tri.fasta&quot;, trploidhet =&gt; refseq)</code></pre><pre><code class="language-none">[ Info: Opening FASTA file at: build-a-yeast-tri.fasta
[ Info: Fabricated 1 chromosomes and written to file</code></pre></div><p>That&#39;s it for the triploid genome.</p><h2><a class="nav-anchor" id="Designing-a-tetraploid-genome-1" href="#Designing-a-tetraploid-genome-1">Designing a tetraploid genome</a></h2><p>Lets make a tetraploid genome, that consists of two diploid genomes, an A genome and a B genome. First I&#39;m going to create a chromosome blueprint with 4 copies:</p><div><pre><code class="language-julia">tetploid = plan_chrom(reflen, 4)</code></pre><pre><code class="language-none">Specification for a chromosome:
 Number of copies: 4
 Length of chromosomes: 230218
 230218bp available in the following stretches:</code></pre></div><p>We want to plan the heterozygosity in this tetraploid such that the number of heterozygous sites within the A and B genomes is ~1%. But we want ~3% of sites to be heterozygous between the A and B genomes.</p><p>First we will say copies 1 + 2 in our blueprint are the A genome, and copies 3 + 4 are the B genome. Next consider heterozygous sites with 2 different alleles, A and B, 6 different patterns are possible:</p><table><tr><th>Copy 1</th><th>Copy 2</th><th>Copy 3</th><th>Copy 4</th></tr><tr><td>A</td><td>A</td><td>B</td><td>B</td></tr><tr><td>B</td><td>B</td><td>A</td><td>A</td></tr><tr><td>A</td><td>B</td><td>A</td><td>B</td></tr><tr><td>B</td><td>A</td><td>B</td><td>A</td></tr><tr><td>A</td><td>B</td><td>B</td><td>A</td></tr><tr><td>B</td><td>A</td><td>A</td><td>B</td></tr></table><p>The first 2 patterns in the table constitute fixed mutations between genomes A and B, but not polymorphism within genomes A and B. This is because Copy 1 and 3 differ, as do copies 2 and 4, yet copies 1 and 2 are the same, as are copies 3 and 4. To achieve a ~3% divergence between genomes A and B, let&#39;s plan some heterozygosity by grouping the copy numbers together:</p><div><pre><code class="language-julia">thet = plan_het(tetploid, 6907, [1, 2], [3, 4])</code></pre><pre><code class="language-none">Specification for a chromosome:
 Number of copies: 4
 Length of chromosomes: 230218
 223311bp available in the following stretches:</code></pre></div><p>By using the groups <code>[1, 2]</code> and <code>[3, 4]</code> we ensure chromosome copy 1 and 2 will have the same allele, as will chromosomes 3 and 4</p><p>The next two patterns in the table constitute polymorphim within the A and B genomes, because copies 1 and 2 differ, as do copies 3 and 4, but copies 1 and 3, and 2 and 4 do not differ.</p><p>The final two patterns in the table constitue both polymorphim within the A and B genomes, and between the A and B genomes, because copies 1 and 2 differ, as do copies 3 and 4, 1 and 3, and 2 and 4.</p><p>One possible way therefore to achieve the desired levels of polymorphism is to Apply a number of these allele patterns, to achieve the 1% and 3% values.</p><div><pre><code class="language-julia">import Random
Random.seed!(1234)
p = suggest_regions(tetploid, 1, 6907)
a1 = suggest_alleles(length(p), [1, 2], [3, 4])
a1 = suggest_alleles(length(p), [1, 1, 2, 2])
thet = plan_het(tetploid, 6907, [1, 2], [3, 4])</code></pre><pre><code class="language-none">Specification for a chromosome:
 Number of copies: 4
 Length of chromosomes: 230218
 223311bp available in the following stretches:</code></pre></div><p><em>This page was generated using <a href="https://github.com/fredrikekre/Literate.jl">Literate.jl</a>.</em></p><footer><hr/><a class="previous" href="../../build-a-genome/"><span class="direction">Previous</span><span class="title">Core concepts &amp; workflow</span></a><a class="next" href="../../sequencing/"><span class="direction">Next</span><span class="title">Core concepts &amp; workflow</span></a></footer></article></body></html>
