<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Reads · Pseudoseq.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link href="../../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><a href="../../index.html"><img class="logo" src="../../assets/logo.png" alt="Pseudoseq.jl logo"/></a><h1>Pseudoseq.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../../search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../../">Home</a></li><li><span class="toctext">Build-a-Genome</span><ul><li><a class="toctext" href="../../build-a-genome/">Core concepts &amp; workflow</a></li><li><span class="toctext">Examples</span><ul><li><a class="toctext" href="../../examples/build-a-yeast/">Build-a-Yeast</a></li></ul></li></ul></li><li><span class="toctext">Sequencing</span><ul><li><a class="toctext" href="../../sequencing/">Core concepts &amp; workflow</a></li><li><span class="toctext">Examples</span><ul><li><a class="toctext" href="../../examples/pe-example/">Paired end reads</a></li><li><a class="toctext" href="../../examples/se-example/">Long single end reads</a></li><li><a class="toctext" href="../../examples/tg-example/">Tagged paired end reads</a></li></ul></li></ul></li><li><span class="toctext">API</span><ul><li><a class="toctext" href="../sequence/">Sequencing</a></li><li><a class="toctext" href="../pool/">Molecule Pool</a></li><li class="current"><a class="toctext" href>Reads</a><ul class="internal"><li><a class="toctext" href="#Exported-functions-1">Exported functions</a></li></ul></li><li><a class="toctext" href="../chromosome-blueprint/">Chromosome Blueprint</a></li></ul></li></ul></nav><article id="docs"><header><nav><ul><li>API</li><li><a href>Reads</a></li></ul><a class="edit-page" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/master/docs/src/api/reads.md"><span class="fa"></span> Edit on GitHub</a></nav><hr/><div id="topbar"><span>Reads</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="API:-Reads-1" href="#API:-Reads-1">API: Reads</a></h1><h2><a class="nav-anchor" id="Exported-functions-1" href="#Exported-functions-1">Exported functions</a></h2><h3><a class="nav-anchor" id="Making-reads-1" href="#Making-reads-1">Making reads</a></h3><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Pseudoseq.make_reads" href="#Pseudoseq.make_reads"><code>Pseudoseq.make_reads</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">make_reads(::Type{PairedEnd}, p::MoleculePool, flen::Int, rlen::Int = flen)</code></pre><p>Create a set of paired-end reads from a pool of DNA molecules <code>p</code>.</p><p><code>flen</code> sets the length of forward read, and <code>rlen</code> sets the length of the reverse read. If you only provide <code>flen</code>, then the function sets <code>rlen = flen</code>.</p><div class="admonition note"><div class="admonition-title">Note</div><div class="admonition-text"><p>If a molecule in the pool is not long enough to create a forward and/or reverse read, then that molecule will simply be skipped. </p></div></div></div></div><a class="source-link" target="_blank" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/d5ba2db025a95293f6851b31a6f7c110d3c268d2/src/reads.jl#L156-L167">source</a><div><div><pre><code class="language-none">make_reads(::Type{SingleEnd}, p::MoleculePool, len::Int)</code></pre><p>Create a set of single-end reads from a pool of DNA molecules <code>p</code>.</p><p><code>len</code> sets the length of the reads.</p><p>The end (strand) from which the reading begins for each DNA molecule in the pool is determined at random for each molecule, with 50:50 probability.</p><p>If you don&#39;t provide a value for <code>len</code>, then the function will read each DNA molecule in it&#39;s entirety.</p><div class="admonition note"><div class="admonition-title">Note</div><div class="admonition-text"><p>If a molecule in the pool is not long enough to create a forward and/or reverse read, then that molecule will simply be skipped. </p></div></div></div></div><a class="source-link" target="_blank" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/d5ba2db025a95293f6851b31a6f7c110d3c268d2/src/reads.jl#L221-L237">source</a><div><div><pre><code class="language-none">make_reads(::Type{TaggedPairs}, p::MoleculePool, flen::Int, rlen::Int = flen)</code></pre><p>Create a set of tagged paired-end reads from a pool of DNA molecules <code>p</code>.</p><p><code>flen</code> sets the length of forward read, and <code>rlen</code> sets the length of the reverse read. If you only provide <code>flen</code>, then the function sets <code>rlen = flen</code>.</p><p>When a set of <code>TaggedPairs</code> is written to file, the tag information is contained in the R1 read of each read-pair. The first 16bp of each R1 read is a sequence that is the tag, and a following 7bp are a buffer between the 16bp tag, and the rest of the read sequence.</p><div class="admonition note"><div class="admonition-title">Note</div><div class="admonition-text"><p>If a molecule in the pool is not long enough to create a forward and/or reverse read, then that molecule will simply be skipped. </p></div></div></div></div><a class="source-link" target="_blank" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/d5ba2db025a95293f6851b31a6f7c110d3c268d2/src/reads.jl#L319-L335">source</a></section><h3><a class="nav-anchor" id="Introducing-errors-1" href="#Introducing-errors-1">Introducing errors</a></h3><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Pseudoseq.mark_errors" href="#Pseudoseq.mark_errors"><code>Pseudoseq.mark_errors</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">mark_errors(reads::Reads, rate::Float64)</code></pre><p>Create a new set of reads, with errors, from an input set of reads.</p><p>When you first create a set of reads using a <a href="#Pseudoseq.make_reads"><code>make_reads</code></a> method, all the reads in that set are perfect reads. In real sequencing experiments, sequencers make errors when reading a DNA molecule, and are characterised by an error rate. This function lets you simulate this characteristic of sequencers, by marking (at random) positions in the reads that are destined to be errors in the output FASTQ.</p><div class="admonition note"><div class="admonition-title">Note</div><div class="admonition-text"><p>Currently, every position in every read is equally likely to be marked as an error.</p></div></div><div class="admonition note"><div class="admonition-title">Note</div><div class="admonition-text"><p>The number of errors introduced into the reads can be calculated.</p><div>\[E = NR\]</div><p>Where <span>$E$</span> is the number of errors, <span>$N$</span> is the total number of bases in your set of reads, and <span>$R$</span> is the error rate you provide to this function.</p></div></div></div></div><a class="source-link" target="_blank" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/d5ba2db025a95293f6851b31a6f7c110d3c268d2/src/reads.jl#L53-L77">source</a></section><h3><a class="nav-anchor" id="Generating-FASTQ-files-1" href="#Generating-FASTQ-files-1">Generating FASTQ files</a></h3><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Pseudoseq.generate" href="#Pseudoseq.generate"><code>Pseudoseq.generate</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">generate(filename::String, reads::Reads)</code></pre><p>Write the <code>reads</code> out to a FASTQ formatted file with the given <code>filename</code>.</p><p>If this method is used with a paired-end read type, then the FASTQ file will be interleaved; all R1 reads will be odd records, and all R2 reads will be even records in the file.</p><div class="admonition note"><div class="admonition-title">Note</div><div class="admonition-text"><p>Reads are named according to the sequence in the input genome they came from. e.g. <code>@Reference_1_R1</code> means the first sequence in the genome, and <code>@Reference_2_R1</code> means the second sequence in the genome.</p></div></div></div></div><a class="source-link" target="_blank" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/d5ba2db025a95293f6851b31a6f7c110d3c268d2/src/reads.jl#L110-L123">source</a><div><div><pre><code class="language-none">generate(R1name::String, R2name::String, reads::Reads{&lt;:PairedReads})</code></pre><p>This method only works for paired reads. Instead of interleaving R1 and R2 reads in a single FASTQ file, R1 and R2 reads are partitioned into two seperate FASTQ files.</p></div></div><a class="source-link" target="_blank" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/d5ba2db025a95293f6851b31a6f7c110d3c268d2/src/reads.jl#L130-L136">source</a></section><footer><hr/><a class="previous" href="../pool/"><span class="direction">Previous</span><span class="title">Molecule Pool</span></a><a class="next" href="../chromosome-blueprint/"><span class="direction">Next</span><span class="title">Chromosome Blueprint</span></a></footer></article></body></html>
