<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Core concepts &amp; workflow · Pseudoseq.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link href="../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><a href="../index.html"><img class="logo" src="../assets/logo.png" alt="Pseudoseq.jl logo"/></a><h1>Pseudoseq.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../">Home</a></li><li><span class="toctext">Build-a-Genome</span><ul><li class="current"><a class="toctext" href>Core concepts &amp; workflow</a><ul class="internal"></ul></li><li><span class="toctext">Examples</span></li></ul></li><li><span class="toctext">Sequencing</span><ul><li><a class="toctext" href="../sequencing/">Core concepts &amp; workflow</a></li><li><span class="toctext">Examples</span><ul><li><a class="toctext" href="../examples/pe-example/">Paired end reads</a></li><li><a class="toctext" href="../examples/se-example/">Long single end reads</a></li><li><a class="toctext" href="../examples/tg-example/">Tagged paired end reads</a></li></ul></li></ul></li><li><span class="toctext">API</span><ul><li><a class="toctext" href="../api/pool/">Molecule Pool</a></li><li><a class="toctext" href="../api/reads/">Reads</a></li><li><a class="toctext" href="../api/chromosome-blueprint/">Chromosome Blueprint</a></li></ul></li></ul></nav><article id="docs"><header><nav><ul><li>Build-a-Genome</li><li><a href>Core concepts &amp; workflow</a></li></ul><a class="edit-page" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/master/docs/src/build-a-genome.md"><span class="fa"></span> Edit on GitHub</a></nav><hr/><div id="topbar"><span>Core concepts &amp; workflow</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="Build-a-Genome:-*Core-concepts-and-basic-workflow*-1" href="#Build-a-Genome:-*Core-concepts-and-basic-workflow*-1">Build-a-Genome: <em>Core concepts &amp; basic workflow</em></a></h1><p><code>Pseudoseq</code> allows you to plan and build genomes and chromosomes that have a certain set of features and peculiarities. The purpose for doing this is not to recreate biology perfectly. The purpose is to create genomes you understand fully (where the repeated content is, which positions are heterozygous and so on).</p><p>Using such genomes can help you both understand and develop an intuition of what current genome assembly tools are doing, and also to help design assembly tools, and perhaps even plan sequencing experiments and form hypotheses.</p><p>This manual includes several examples showing how to genomes with certain characteristics. But the core workflow, and important concepts are explained below.</p><h3><a class="nav-anchor" id=".-Creating-chromosome-blueprints-1" href="#.-Creating-chromosome-blueprints-1">1. Creating chromosome blueprints</a></h3><p>Chromosome blueprints are the backbone of simulating genomes with Pseudoseq. Chromosome blueprints determine what the nature of one chromosome in a genome. Depending on the genome, any given chromosome may be present in multiple copies. Diploids for example have two copies of every chromosome.</p><p>The first step in simulating any artificial genome is to create one or more blank chromosome blueprints. The <a href="../api/chromosome-blueprint/#Pseudoseq.plan_chrom"><code>plan_chrom</code></a> function is used for this.</p><p>For example, this:</p><pre><code class="language-julia-repl">julia&gt; using Pseudoseq

julia&gt; c = plan_chrom(100, 2)
Specification for a chromosome:
 Number of copies: 2
 Length of chromosomes: 100
 100bp available in the following stretches:</code></pre><p>will create a blank blueprint for 2 copies of a chromosome of 100bp length.</p><h3><a class="nav-anchor" id=".-Adding-planned-features-to-a-chromosome-blueprint-1" href="#.-Adding-planned-features-to-a-chromosome-blueprint-1">2. Adding planned features to a chromosome blueprint</a></h3><p>Once you have one or more chromosome blueprints, you can add features to them.</p><p>Currently the supported features, and the functions that add to a blueprint include:</p><ul><li>Repetitions: <a href="../api/chromosome-blueprint/#Pseudoseq.plan_repetition"><code>plan_repetition</code></a></li><li>Heterozygosity: <a href="../api/chromosome-blueprint/#Pseudoseq.plan_het"><code>plan_het</code></a></li></ul><div class="admonition note"><div class="admonition-title">Note</div><div class="admonition-text"><p>Every time one of these <code>plan_*</code> functions is used to add a feature to a chromosome blueprint, a new chromosome blueprint is created.</p></div></div><div class="admonition note"><div class="admonition-title">Note</div><div class="admonition-text"><p>If a region of a chromosome is used to plan some feature. E.g. position 5 is used to create some heterozygosity, that same position cannot be used for a second feature, it is consumed and cannot be used again.</p></div></div><p>For example, below will make it so as about 50% of the two chromosome copies  are heterozygous:</p><pre><code class="language-julia-repl">julia&gt; chet = plan_het(c, 50, 2)
Specification for a chromosome:
 Number of copies: 2
 Length of chromosomes: 100
 50bp available in the following stretches:</code></pre><p>There is a utility function <a href="../api/chromosome-blueprint/#Pseudoseq.suggest_regions"><code>suggest_regions</code></a> available to help plan features. Say you wanted to see where you could place 3 5bp repetitions, you could do the following:</p><pre><code class="language-julia-repl">julia&gt; r = suggest_regions(chet, 5, 3)
3-element Array{UnitRange{Int64},1}:
 56:60
 62:66
 37:41</code></pre><h3><a class="nav-anchor" id=".-Fabricate-a-FASTA-from-chromosome-blueprints-1" href="#.-Fabricate-a-FASTA-from-chromosome-blueprints-1">3. Fabricate a FASTA from chromosome blueprints</a></h3><p>Once you have a set of chromosome blueprints with the features planned that you desire, you can fabricate a FASTA file containing the sequences of these chromosomes. To do that, use the <a href="@ref"><code>fabricate</code></a> method that accepts a filename, and a variable number of chromosome blueprints...</p><pre><code class="language-julia-repl">julia&gt; fabricate(&quot;mychrom.fasta&quot;, chet)
[ Info: Opening FASTA file at: mychrom.fasta
[ Info: Fabricated 1 chromosomes and written to file</code></pre><footer><hr/><a class="previous" href="../"><span class="direction">Previous</span><span class="title">Home</span></a><a class="next" href="../sequencing/"><span class="direction">Next</span><span class="title">Core concepts &amp; workflow</span></a></footer></article></body></html>
