<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Core concepts &amp; workflow · Pseudoseq.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link href="../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><a href="../index.html"><img class="logo" src="../assets/logo.png" alt="Pseudoseq.jl logo"/></a><h1>Pseudoseq.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../">Home</a></li><li><span class="toctext">Sequencing</span><ul><li><a class="toctext" href="../sequencing/">Core concepts &amp; workflow</a></li><li><span class="toctext">Examples</span><ul><li><a class="toctext" href="../examples/sequencing/pe-example/">Paired end reads</a></li><li><a class="toctext" href="../examples/sequencing/se-example/">Long single end reads</a></li><li><a class="toctext" href="../examples/sequencing/tg-example/">Tagged paired end reads</a></li></ul></li></ul></li><li><span class="toctext">Build-a-Genome</span><ul><li class="current"><a class="toctext" href>Core concepts &amp; workflow</a><ul class="internal"><li><a class="toctext" href="#.-Creating-chromosome-blueprints-1">1. Creating chromosome blueprints</a></li><li><a class="toctext" href="#.-Adding-planned-features-to-a-chromosome-blueprint-1">2. Adding planned features to a chromosome blueprint</a></li><li><a class="toctext" href="#.-Fabricate-a-FASTA-from-chromosome-blueprints-1">3. Fabricate a FASTA from chromosome blueprints</a></li></ul></li><li><span class="toctext">Examples</span><ul><li><a class="toctext" href="../examples/build-a-genome/build-a-yeast/">Build-a-Yeast</a></li></ul></li></ul></li><li><span class="toctext">API</span><ul><li><span class="toctext">Sequencing</span><ul><li><a class="toctext" href="../api/sequence/">sequence</a></li><li><a class="toctext" href="../api/pool/">Molecule Pool</a></li><li><a class="toctext" href="../api/reads/">Reads</a></li></ul></li><li><a class="toctext" href="../api/chromosome-blueprint/">Build-a-Genome</a></li></ul></li></ul></nav><article id="docs"><header><nav><ul><li>Build-a-Genome</li><li><a href>Core concepts &amp; workflow</a></li></ul><a class="edit-page" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/master/docs/src/build-a-genome.md"><span class="fa"></span> Edit on GitHub</a></nav><hr/><div id="topbar"><span>Core concepts &amp; workflow</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="Build-a-Genome:-*Core-concepts-and-basic-workflow*-1" href="#Build-a-Genome:-*Core-concepts-and-basic-workflow*-1">Build-a-Genome: <em>Core concepts &amp; basic workflow</em></a></h1><p><code>Pseudoseq</code> allows you to plan and build genomes and chromosomes that have a certain set of features and peculiarities. The purpose for doing this is not to recreate biology perfectly. The purpose is to create genomes you understand fully (where the repeated content is, which positions are heterozygous and so on).</p><p>Using such genomes can help you both understand and develop an intuition of what current genome assembly tools are doing, and also to help design assembly tools, and perhaps even plan sequencing experiments and form hypotheses.</p><p>This manual includes several examples showing how to plan genomes with certain characteristics. But the core workflow, and important concepts are explained below, in 3 steps.</p><h2><a class="nav-anchor" id=".-Creating-chromosome-blueprints-1" href="#.-Creating-chromosome-blueprints-1">1. Creating chromosome blueprints</a></h2><p>Chromosome blueprints are the backbone of simulating genomes with Pseudoseq.</p><p>Chromosome blueprints determine the nature of one chromosome in a genome.</p><p>You can think of chromosome blueprints in Pseudoseq as: <em>a collection of operations which, when applied to some seed sequence, result in a set of N homologous sequences</em>.</p><div class="admonition note"><div class="admonition-title">Note</div><div class="admonition-text"><p>Chromosome blueprints are <em>immutable</em>: Adding an operation to a chromosome blueprint creates a new blueprint, which is a copy of the old blueprint with the addition of the new operation.</p></div></div><div class="admonition note"><div class="admonition-title">Note</div><div class="admonition-text"><p>Ideally, for any chromosome blueprint you construct with Pseudoseq, for each operation is possesses, it must be possible (at least in principle) to be able to intuit what the effect of that operation will be on:</p><ol><li>The Khmer distribution produced by sequencing reads of the fabricated chromosome.</li><li>The structure of a sequence graph produced by sequencing reads of the fabricated chromosome.</li></ol><p>Therefore, we have made the design decision that <em>no two operations in a chromosome blueprint may affect the same position(s) of the genome in a conflicting manner.</em> To meet this requirement, certain operations &quot;consume&quot; a region of the chromosome planned in a blueprint. If a region is consumed, another operation that would also affect that region cannot be added to the blueprint.</p></div></div><p>Depending on the genome, any given chromosome may be present in multiple copies. Diploids, for example have two copies of every chromosome.</p><p>The first step in simulating any artificial genome is to create one or more blank chromosome blueprints. The <a href="../api/chromosome-blueprint/#Pseudoseq.plan_chrom"><code>plan_chrom</code></a> function is used for this.</p><p>For example, this:</p><pre><code class="language-julia-repl">julia&gt; using Pseudoseq

julia&gt; c = plan_chrom(100, 2)
Specification for a chromosome:
 Number of copies: 2
 Length of chromosomes: 100
 100bp available in the following stretches:</code></pre><p>will create a blank blueprint for 2 copies of a chromosome of 100bp length. From the output you can see that it prints for you the number of chromosome copies, the length of the chromosomes, and a list of available, <em>unconsumed</em> regions of the chromosome (see note above).</p><h2><a class="nav-anchor" id=".-Adding-planned-features-to-a-chromosome-blueprint-1" href="#.-Adding-planned-features-to-a-chromosome-blueprint-1">2. Adding planned features to a chromosome blueprint</a></h2><p>After creating a fresh chromosome blueprint, no plans (operations) have been added yet.</p><p>If you were to <a href="../api/chromosome-blueprint/#Pseudoseq.fabricate"><code>fabricate</code></a> this blank blueprint, you would get <code>N</code> identical DNA sequences as output, where <code>N</code> is the number of copies the blueprint was planning.</p><p>Once you have one or more chromosome blueprints, you can add features to them.</p><p>This is done with a series of consistently named <code>plan_*</code> functions.</p><div class="admonition note"><div class="admonition-title">Note</div><div class="admonition-text"><p>Remember; blueprints are immutable, so every time one of these <code>plan_*</code> functions is used to add a feature to a chromosome blueprint, a new chromosome blueprint is created.</p></div></div><h3><a class="nav-anchor" id="Repetitions-1" href="#Repetitions-1">Repetitions</a></h3><p>A repetition is a segment of a sequence, that has the exact same motif, as  another segment of the sequence.</p><p>In Pseudoseq, to plan a repetition, you specify a region the repetition will copy, sometimes called the <code>from</code> region. You also specify a region where the motif in the <code>from</code> region will be replicated, called the <code>to</code> region. So you might find it helpful to imagine a planned repetition as a kind of copy-paste operation that occurs during <a href="../api/chromosome-blueprint/#Pseudoseq.fabricate"><code>fabricate</code></a>.</p><div class="admonition note"><div class="admonition-title">Note</div><div class="admonition-text"><p>Repetitions consume the <code>to</code> region of the chromosome blueprint to which they are added. Repetitions <em>do not</em> consume the <code>from</code> region, so other operations are free to affect the motif in the <code>from</code> region. Just remember that the repetition will replicate anything in the <code>from</code> region, including other features such as heterozygosity that occur in <code>from</code>.</p></div></div><p>Repetitions are added to a genome blueprint using the <a href="../api/chromosome-blueprint/#Pseudoseq.plan_repetition"><code>plan_repetition</code></a> function.</p><h3><a class="nav-anchor" id="Heterozygosity-1" href="#Heterozygosity-1">Heterozygosity</a></h3><p>A heterozygosity describes a base position at which the copies of the chromosome in the blueprint differ.</p><p>For a blueprint with 2 copies, both copies will differ at a given position.</p><p>For a triploid, at a heterozygous position, all 3 copies might differ from each other. Alternatively, it is possible that 2 copies are the same, but they differ from a 3rd copy. This applies for blueprints with higher copy numbers too: at a heterozygous position some copies will differ from each other at that position, but some of the copies might be the same.</p><div class="admonition note"><div class="admonition-title">Note</div><div class="admonition-text"><p>Heterozygosity operations consume the position at which they are defined.</p></div></div><p>Heterozygous positions are planned using the <a href="../api/chromosome-blueprint/#Pseudoseq.plan_het"><code>plan_het</code></a> function.</p><p>For example, below will make it so as about 50% of the two chromosome copies  are heterozygous:</p><pre><code class="language-julia-repl">julia&gt; using Pseudoseq

julia&gt; c = plan_chrom(100, 2)
Specification for a chromosome:
 Number of copies: 2
 Length of chromosomes: 100
 100bp available in the following stretches:

julia&gt; chet = plan_het(c, .50, 2)
Specification for a chromosome:
 Number of copies: 2
 Length of chromosomes: 100
 50bp available in the following stretches:</code></pre><p>The above use of <a href="../api/chromosome-blueprint/#Pseudoseq.plan_het"><code>plan_het</code></a> allows the function to choose which sites in the blueprint are heterozygous, and how to allocate the bases at the heterozygous sites. We only instruct the function that 50% of the genome should be heterozygous, and that there should be 2 possible bases at each position (the only option really: the blueprint plans for a diploid - 2 chromosome copies).</p><p>You can also take more fine-grained control:</p><pre><code class="language-julia-repl">julia&gt; using Pseudoseq, BioSequences

julia&gt; c = plan_chrom(100, 2)
Specification for a chromosome:
 Number of copies: 2
 Length of chromosomes: 100
 100bp available in the following stretches:

julia&gt; chet = plan_het(c, 20, [DNA_T, DNA_A])
Specification for a chromosome:
 Number of copies: 2
 Length of chromosomes: 100
 99bp available in the following stretches:</code></pre><p>Here I planned a heterozygous position, at one site in the chromosome (20), and I set the state of the first copy to <code>DNA_T</code>, and the state of the second copy to <code>DNA_A</code>.</p><p>So as you can see, <a href="../api/chromosome-blueprint/#Pseudoseq.plan_het"><code>plan_het</code></a> is very flexible. Check it&#39;s API documentation, and the &quot;Build-A-Yeast&quot; example walkthrough to see more examples of <a href="../api/chromosome-blueprint/#Pseudoseq.plan_het"><code>plan_het</code></a> use.</p><h3><a class="nav-anchor" id="Utility-functions-1" href="#Utility-functions-1">Utility functions</a></h3><p>There is a utility function <a href="../api/chromosome-blueprint/#Pseudoseq.suggest_regions"><code>suggest_regions</code></a> available to help you plan where to place features.</p><p>Say you wanted to see where you could place 3 5bp repetitions, you could do the following:</p><pre><code class="language-julia-repl">julia&gt; r = suggest_regions(chet, 5, 6)
6-element Array{UnitRange{Int64},1}:
 15:19
 3:7
 9:13
 95:99
 64:68
 56:60</code></pre><p>So now you have 6 5bp regions, every 2 regions defining a <code>from</code> and a <code>to</code> region for a single repetition. </p><p>You can provide <code>r</code> as an input to <a href="../api/chromosome-blueprint/#Pseudoseq.plan_repetition"><code>plan_repetition</code></a>.</p><pre><code class="language-julia-repl">julia&gt; crep = plan_repetition(chet, r)
Specification for a chromosome:
 Number of copies: 2
 Length of chromosomes: 100
 84bp available in the following stretches:</code></pre><p>Another utility function <a href="../api/chromosome-blueprint/#Pseudoseq.suggest_alleles"><code>suggest_alleles</code></a> is available to help you plan nucleotide patterns at heterozygous sites.</p><p>Say you wanted to plan a pattern in which two of three chromosome copies had the same base, and a third differed.</p><pre><code class="language-julia-repl">julia&gt; suggest_alleles(3, 2)
3-element Array{BioSymbols.DNA,1}:
 DNA_G
 DNA_G
 DNA_A</code></pre><p>The above asks for an allele pattern for 3 copies of a chromosome, with 2 possible alleles, which chromosome copy gets which allele is random.</p><p>You can also take more control and specify which chromosome copy gets which state:</p><pre><code class="language-julia-repl">julia&gt; suggest_alleles(chet, [1, 1, 2])
ERROR: MethodError: no method matching suggest_alleles(::Pseudoseq.ChromosomeBlueprint, ::Array{Int64,1})
Closest candidates are:
  suggest_alleles(!Matched::Array{Int64,1}...) at /home/travis/build/bioinfologics/Pseudoseq.jl/src/build_a_genome/plan/polymorphism.jl:98
  suggest_alleles(!Matched::Int64, ::Any...) at /home/travis/build/bioinfologics/Pseudoseq.jl/src/build_a_genome/plan/polymorphism.jl:140

julia&gt; # Or alternatively put:
       suggest_alleles(chet, [1, 2], [3])
ERROR: MethodError: no method matching suggest_alleles(::Pseudoseq.ChromosomeBlueprint, ::Array{Int64,1}, ::Array{Int64,1})
Closest candidates are:
  suggest_alleles(!Matched::Array{Int64,1}...) at /home/travis/build/bioinfologics/Pseudoseq.jl/src/build_a_genome/plan/polymorphism.jl:98
  suggest_alleles(!Matched::Int64, ::Any...) at /home/travis/build/bioinfologics/Pseudoseq.jl/src/build_a_genome/plan/polymorphism.jl:140</code></pre><p>In the above, which chromosomes get the same allele is user-specified, but which bases those alleles are, is randomly determined.</p><p>You can get many suggestions at once:</p><pre><code class="language-julia-repl">julia&gt; suggest_alleles(5, [1, 1, 2])
5-element Array{Array{BioSymbols.DNA,1},1}:
 [DNA_G, DNA_G, DNA_A]
 [DNA_T, DNA_T, DNA_C]
 [DNA_G, DNA_G, DNA_C]
 [DNA_T, DNA_T, DNA_G]
 [DNA_C, DNA_C, DNA_G]</code></pre><p>See the API docs for <a href="../api/chromosome-blueprint/#Pseudoseq.suggest_alleles"><code>suggest_alleles</code></a> for more details on the arguments permitted by the different methods. </p><h2><a class="nav-anchor" id=".-Fabricate-a-FASTA-from-chromosome-blueprints-1" href="#.-Fabricate-a-FASTA-from-chromosome-blueprints-1">3. Fabricate a FASTA from chromosome blueprints</a></h2><p>Once you have a set of chromosome blueprints with the features planned that you desire, you can fabricate the sequences of these chromosomes.</p><p>To do that, use the <a href="../api/chromosome-blueprint/#Pseudoseq.fabricate"><code>fabricate</code></a> method.</p><p>You can simply fabricate the sequences, for use in the interactive session:</p><pre><code class="language-julia-repl">julia&gt; fabricate(chet)
2-element Array{BioSequences.BioSequence{BioSequences.DNAAlphabet{2}},1}:
 ACGCAGGGGTGGTTTCACCTAGGTTGACGGTGGGTAATT…GAATTGTTTTACCTTATCGTATGGTATGGGCGCGCAGGT
 ACGCAGGGGTGGTTTCACCAAGGTTGACGGTGGGTAATT…GAATTGTTTTACCTTATCGTATGGTATGGGCGCGCAGGT</code></pre><p>Or have the sequences output to a FASTA formatted file.</p><pre><code class="language-julia-repl">julia&gt; fabricate(&quot;mychrom.fasta&quot;, chet)
[ Info: Opening FASTA file at: mychrom.fasta
[ Info: Fabricated 1 chromosomes and written to file</code></pre><p>A randomly generated seed sequence is used to start the fabrication process unless you provide one. See the API docs for <a href="../api/chromosome-blueprint/#Pseudoseq.fabricate"><code>fabricate</code></a> for more details.</p><p>And that&#39;s all there is to building a chromosome with Pseudoseq. To build a genome with more than one chromosome, simply build a set of chromosome blueprints.</p><p>Now check out the examples to see how various genomes can be built with Pseudoseq.</p><footer><hr/><a class="previous" href="../examples/sequencing/tg-example/"><span class="direction">Previous</span><span class="title">Tagged paired end reads</span></a><a class="next" href="../examples/build-a-genome/build-a-yeast/"><span class="direction">Next</span><span class="title">Build-a-Yeast</span></a></footer></article></body></html>
