<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Build-a-Genome · Pseudoseq.jl</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img src="../../assets/logo.png" alt="Pseudoseq.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit">Pseudoseq.jl</span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Manual</span><ul><li><input class="collapse-toggle" id="menuitem-2-1" type="checkbox"/><label class="tocitem" for="menuitem-2-1"><span class="docs-label">Sequencing</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../man/sequencing/concepts/">Core concepts &amp; workflow</a></li><li><input class="collapse-toggle" id="menuitem-2-1-2" type="checkbox"/><label class="tocitem" for="menuitem-2-1-2"><span class="docs-label">Examples</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../man/sequencing/examples/pe-example/">Paired end reads</a></li><li><a class="tocitem" href="../../man/sequencing/examples/se-example/">Long single end reads</a></li><li><a class="tocitem" href="../../man/sequencing/examples/tg-example/">Tagged paired end reads</a></li></ul></li></ul></li><li><input class="collapse-toggle" id="menuitem-2-2" type="checkbox"/><label class="tocitem" for="menuitem-2-2"><span class="docs-label">Build-a-Genome</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../man/build-a-genome/concepts/">Core concepts &amp; workflow</a></li><li><input class="collapse-toggle" id="menuitem-2-2-2" type="checkbox"/><label class="tocitem" for="menuitem-2-2-2"><span class="docs-label">Examples</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../man/build-a-genome/examples/build-a-yeast/">Build-a-Yeast</a></li></ul></li></ul></li></ul></li><li><span class="tocitem">API</span><ul><li><input class="collapse-toggle" id="menuitem-3-1" type="checkbox"/><label class="tocitem" for="menuitem-3-1"><span class="docs-label">Sequencing</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../sequence/">sequence</a></li><li><a class="tocitem" href="../pool/">Molecule Pool</a></li><li><a class="tocitem" href="../reads/">Reads</a></li></ul></li><li class="is-active"><a class="tocitem" href>Build-a-Genome</a><ul class="internal"><li><a class="tocitem" href="#Exported-functions-1"><span>Exported functions</span></a></li></ul></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">API</a></li><li class="is-active"><a href>Build-a-Genome</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Build-a-Genome</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/master/docs/src/api/chromosome-blueprint.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="API:-Build-a-Genome-1"><a class="docs-heading-anchor" href="#API:-Build-a-Genome-1">API: Build-a-Genome</a><a class="docs-heading-anchor-permalink" href="#API:-Build-a-Genome-1" title="Permalink"></a></h1><h2 id="Exported-functions-1"><a class="docs-heading-anchor" href="#Exported-functions-1">Exported functions</a><a class="docs-heading-anchor-permalink" href="#Exported-functions-1" title="Permalink"></a></h2><h3 id="Making-an-empty-chromosome-blueprint-1"><a class="docs-heading-anchor" href="#Making-an-empty-chromosome-blueprint-1">Making an empty chromosome blueprint</a><a class="docs-heading-anchor-permalink" href="#Making-an-empty-chromosome-blueprint-1" title="Permalink"></a></h3><article class="docstring"><header><a class="docstring-binding" id="Pseudoseq.plan_chrom" href="#Pseudoseq.plan_chrom"><code>Pseudoseq.plan_chrom</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Create an empty blueprint for <code>n</code> copies of a chromosome of <code>len</code> base pairs.</p></div></section></article><h3 id="Planning-motif-repetitions-along-chromosomes-1"><a class="docs-heading-anchor" href="#Planning-motif-repetitions-along-chromosomes-1">Planning motif repetitions along chromosomes</a><a class="docs-heading-anchor-permalink" href="#Planning-motif-repetitions-along-chromosomes-1" title="Permalink"></a></h3><article class="docstring"><header><a class="docstring-binding" id="Pseudoseq.plan_repetition" href="#Pseudoseq.plan_repetition"><code>Pseudoseq.plan_repetition</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">plan_repetition(cb::ChromosomeBlueprint, from::UnitRange{Int}, to::UnitRange{Int})</code></pre><p>Plan a repetition in a chromosome, where the bases in the <code>from</code> region of the chromosome, are guaranteed to occur again in the <code>to</code> region of the chromosome.</p><p>Every copy of the chromosome will have this same repetition.</p><p>Creates a new chromosome blueprint, based on the input blueprint <code>cb</code>.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>In the new blueprint, the <code>to</code> region of the planned chromosome will be consumed, and cannot be used to plan any other subsequently added features.</p></div></div></div></section><section><div><pre><code class="language-none">plan_repetition(cb::ChromosomeBlueprint, from::Int, to::Int, size::Int)</code></pre><p>Plan a repetition in a chromosome, where the bases in the <code>from</code>:(<code>from</code> + <code>size</code> - 1) region of the chromosome, are guaranteed to occur again in the <code>to</code>:(<code>to</code> + <code>size</code> - 1) region of the chromosome.</p><p>Every copy of the chromosome will have this same repetition.</p><p>Creates a new chromosome blueprint, based on the input blueprint <code>cb</code>.</p></div></section><section><div><pre><code class="language-none">plan_repetition(cb::ChromosomeBlueprint, intervals::Vector{UnitRange{Int}})</code></pre><p>A conveinience method of <code>plan_repetition</code>. Designed to ease the process of planing a series of repetitions in a chromosome.</p><p>Every copy of the chromosome will have these same repetitions.</p><p>Creates a new chromosome blueprint, based on the input blueprint <code>cb</code>.</p><div class="admonition is-success"><header class="admonition-header">Tip</header><div class="admonition-body"><p>Use the <a href="#Pseudoseq.suggest_regions"><code>suggest_regions</code></a> function to help decide on a set of sites to make heterozygous.</p></div></div><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>The number of intervals provided must be an even number. This is because intervals 1 &amp; 2  define the first repeat, intervals 3 &amp; 4 define the second, and so on.</p></div></div><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>In the new blueprint, the regions the repetitions occupy, have been consumed, and cannot be used to plan any other subsequently added features.</p></div></div></div></section></article><h3 id="Planning-heterozygosity-between-chromosome-copies-1"><a class="docs-heading-anchor" href="#Planning-heterozygosity-between-chromosome-copies-1">Planning heterozygosity between chromosome copies</a><a class="docs-heading-anchor-permalink" href="#Planning-heterozygosity-between-chromosome-copies-1" title="Permalink"></a></h3><article class="docstring"><header><a class="docstring-binding" id="Pseudoseq.plan_het" href="#Pseudoseq.plan_het"><code>Pseudoseq.plan_het</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">plan_het(cb::ChromosomeBlueprint, pos::Int, alleles::Vector{DNA})</code></pre><p>Plan heterozygosity between copies of a chromosome at position <code>pos</code>.</p><p>The <code>alleles</code> vector must contain a nucleotide for each copy of the chromosome.</p><p>For example if you provided a vector of <code>[DNA_A, DNA_C]</code>, for the heterozygous site you&#39;re defining at <code>pos</code>, the first copy of the chromosome will have an <code>A</code>, and the second copy of the chromosome will have a <code>C</code>.</p><p>Creates a new chromosome blueprint, based on the input blueprint <code>cb</code>.</p><div class="admonition is-success"><header class="admonition-header">Tip</header><div class="admonition-body"><p>Use the <a href="#Pseudoseq.suggest_alleles"><code>suggest_alleles</code></a> function to help decide on a set of alleles to use.</p></div></div><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>In the new blueprint, the positions that were used to plan the heterozygosity, have been consumed, and cannot be used to plan any other subsequently added features.</p></div></div></div></section><section><div><pre><code class="language-none">plan_het(cb::ChromosomeBlueprint, pos, alle...)</code></pre><p>A generic method of <code>plan_het</code>.</p><p>Creates a new chromosome blueprint, based on the input blueprint <code>cb</code>.</p><p>The <code>pos</code> argument determines which sites in a chromosome are made heterozygous, and depending on the type of argument provided, the behaviour differs slightly:</p><ul><li><p>If <code>pos</code> is an integer, then that many available sites are selected at random to be heterozygous.</p></li><li><p>If <code>pos</code> is a float, then it is treated as a proportion of the length of the, chromosome, which is converted into an integer, and that many sites are selected at random to be heterozygous.</p></li><li><p>If <code>pos</code> is a vector of integers, it is trated as a list of positions the user wants to be heterozygous.</p></li></ul><p>The <code>alle...</code> argument is a vararg argument which defines the how bases will be  allocated at each heterozygous site.</p><p>Depending on the type of the argument provided, the behaviour differs slightly, based on the different methods of <a href="#Pseudoseq.suggest_alleles"><code>suggest_alleles</code></a>:</p><ul><li><p>If <code>alle...</code> is a single integer value. It is taken to mean the number of states each heterozygous site has. At a minumum this number must be two, as a site with only one state is not heterozygous by definition. For each site, which chromosome copies get which state is determined at random.</p></li><li><p>If <code>alle...</code> is a vector of integers (one integer for every chromosome copy), then the nth value in the vector dictates the group the nth chromosome copy belongs to. For example for a triploid, the vector <code>[2, 1, 2]</code> means that there are two groups (group 1 and group 2), and that the second chromosome copy of the three belongs to group 1, and the other two copies belong to group</p><ol><li>Chromosome copies that belong to the same group will get the same base at</li></ol><p>each heterozygous site. In our example, copies 1 and 3 will get the same base, as they have been given the same group number. Chromosome copy 2 will end up with a different base. Which base corresponds to which group number will be determined at random.</p></li><li><p>If <code>alle...</code> is multiple vectors of integers, then each vector constitutes a group, and each value in the vector determines which copies are in that group. It is alternative way to giving the same information as you can with a single vector (as above). E.g. consider the example of <code>[2, 1, 2]</code> from the previous point. This can be expressed as two vectors of <code>[2] and [1, 3]</code>: The first vector contains a number indicating the second chromosome copy, and the second vector contains numbers representing the first and third chromosome copy.</p></li><li><p>If <code>alle...</code> is a vector of nucleotide vectors, then the nth vector of nucleotides determines which chromosome copy recieves which base at the nth heterozygous position. For example, for a triploid, <code>[[DNA_A, DNA_A, DNA_T], [DNA_G, DNA_C, DNA_C]]</code> would mean that at the first heterozygous site, the first and second copies of the chromosome would recieve an <code>DNA_A</code> base, and the third copy would recieve a <code>DNA_T</code> base. For the second heterozygous position, the first copy of the chromosome would recieve a DNA<em>G base, and the other two copies will recieve a `DNA</em>C` base.</p></li></ul><p>(See also: <a href="#Pseudoseq.plan_repetition"><code>plan_repetition</code></a>)</p></div></section></article><h3 id="Fabricate-1"><a class="docs-heading-anchor" href="#Fabricate-1">Fabricate</a><a class="docs-heading-anchor-permalink" href="#Fabricate-1" title="Permalink"></a></h3><article class="docstring"><header><a class="docstring-binding" id="Pseudoseq.fabricate" href="#Pseudoseq.fabricate"><code>Pseudoseq.fabricate</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">fabricate(file::String, cb...)</code></pre><p>Fabricate the sequence(s) planned in a number chromosome blueprints.</p><p>The fabricated sequences will be written out to the FASTA formatted file <code>file</code>.</p><p><code>cb...</code> should be provided as a series of blueprint, seed-sequence pairs. E.g. <code>fabricate(&quot;mygenome.fasta&quot;, chr1plan =&gt; ch1seq, ch2plan =&gt; ch2seq)</code>.</p></div></section><section><div><pre><code class="language-none">fabricate(fw::FASTA.Writer, cb...)</code></pre><p>Fabricate the sequence(s) planned in a number chromosome blueprints.</p><p>The fabricated sequences will be written out to the FASTA formatted file <code>fw</code>.</p><p><code>cb...</code> should be provided as a series of blueprint, seed-sequence pairs. E.g. <code>fabricate(&quot;mygenome.fasta&quot;, chr1plan =&gt; ch1seq, ch2plan =&gt; ch2seq)</code>.</p></div></section><section><div><pre><code class="language-none">fabricate(cb::ChromosomeBlueprint)</code></pre><p>Fabricate the sequence(s) planned in the chromosome blueprint.</p><p>A random DNA sequence will be generated to use as a seed sequence.</p><p>A sequence will be built for each chromosome copy in the blueprint.</p></div></section><section><div><pre><code class="language-none">fabricate(cb::ChromosomeBlueprint, seed::LongSequence{DNAAlphabet{2}})</code></pre><p>Fabricate a DNA sequence by applying the planned features in a chromosome blueprint, to some initial starting seed sequence.</p><p>A sequence will be built for each chromosome copy in the blueprint.</p></div></section></article><h3 id="Utility-functions-1"><a class="docs-heading-anchor" href="#Utility-functions-1">Utility functions</a><a class="docs-heading-anchor-permalink" href="#Utility-functions-1" title="Permalink"></a></h3><article class="docstring"><header><a class="docstring-binding" id="Pseudoseq.suggest_regions" href="#Pseudoseq.suggest_regions"><code>Pseudoseq.suggest_regions</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">suggest_regions(cp::ChromosomeBlueprint, size::Int, n::Int)</code></pre><p>A useful utility function to assist planning features in a chromosome blueprint.</p><p>This method returns a vector of non-overlapping, regions of the chromosome planned represented by the ChromosomeBlueprint <code>cb</code>.</p><p>These regions are free regions: they are untouched by any other planned features, and so may be used when planning other features (See also: <a href="#Pseudoseq.plan_repetition"><code>plan_repetition</code></a>, <a href="#Pseudoseq.plan_het"><code>plan_het</code></a>).</p><p>The regions will be <code>size</code>bp in length.</p><div class="admonition is-warning"><header class="admonition-header">Warning</header><div class="admonition-body"><p>This function was designed for use interactively in a julia session.</p><p>If this method cannot find <code>n</code> free regions of the size you&#39;ve asked for, it will still give an output vector containing the regions it did manage to find, but it will issue a warning to the terminal. This will get increasingly likely as you fill the chromosome blueprint up with features.</p><p>If you are using this method in a script or program where you depend on a reliable number of regions, either add a check to make sure you got the number of regions you need, or use this method interactively, and hard code an appropriate output into your script.</p></div></div></div></section></article><article class="docstring"><header><a class="docstring-binding" id="Pseudoseq.suggest_alleles" href="#Pseudoseq.suggest_alleles"><code>Pseudoseq.suggest_alleles</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">suggest_alleles(groups::Vector{Int})</code></pre><p>Suggest an allele pattern for a single heterozygous site.</p><p>The <code>groups</code> vector should have a number of  The nth value in the <code>groups</code> designates a group to the nth chromosome copy.</p><p>E.g. consider for a triploid, the vector <code>[2, 1, 2]</code> means that there are two groups (group 1 and group 2), and that the second chromosome copy of the three belongs to group 1, and the other two copies belong to group 2.</p><p>Chromosome copies that belong to the same group will get the same base at each heterozygous site.</p><p>For the vector <code>[2, 1, 2]</code>, this method might return a suggested allele pattern such as <code>[DNA_A, DNA_G, DNA_A]</code>.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>Which base corresponds to which group number will be determined at random.</p></div></div></div></section><section><div><pre><code class="language-none">suggest_alleles(groups::Vector{Int}...)</code></pre><p>Suggest an allele pattern for a single heterozygous site.</p><p>Multiple integer vectors make up <code>groups</code>.</p><p>Each vector defines a group, and each value in a vector determines which copies are allocated to that group.</p><p>E.g. consider for a triploid, two vectors of <code>[2] and [1, 3]</code> passed as arguments to <code>groups</code>. They define two groups, the first vector contains a number denoting the second chromosome copy, and he second vector contains numbers representing the first and third chromosome copy.</p><p>For the vectors <code>[2]</code> and <code>[1, 3]</code>, this method might return a suggested allele pattern such as <code>[DNA_A, DNA_G, DNA_A]</code>.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>Which base corresponds to which group number will be determined at random.</p></div></div></div></section><section><div><pre><code class="language-none">suggest_alleles(ncopies::Int, ngroups::Int)</code></pre><p>Suggest an allele pattern for a single heterozygous site.</p><p>By providing a number representing the number of copies of a chromosome, and a number representing the number of groups (which at a minimum must be 2). This method of <code>suggest_alleles</code> will randomly allocate the <code>ncopies</code> of the chromosome into the <code>ngroups</code> groups.</p><p>For example, if you ask for a suggested allele pattern for 3 copies of a chromosome, and 2 groups, this method might return a suggested pattern of <code>[DNA_G, DNA_A, DNA_G]</code>.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>Which base corresponds to which group number will be determined at random.</p></div></div></div></section><section><div><pre><code class="language-none">suggest_alleles(npositions::Int, args...)</code></pre><p>Suggest allele patterns for numerous heterozygous sites.</p><p>This method of <code>suggest_alleles</code> calls another <code>suggest_alleles</code> with <code>args...</code> <code>npositions</code> times, and returns a vector of the results of each call.</p></div></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../reads/">« Reads</a></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Tuesday 10 December 2019 23:07">Tuesday 10 December 2019</span>. Using Julia version 1.2.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
