<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Paired end reads · Pseudoseq.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link href="../../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><a href="../../index.html"><img class="logo" src="../../assets/logo.png" alt="Pseudoseq.jl logo"/></a><h1>Pseudoseq.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../../search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../../">Home</a></li><li><span class="toctext">Build-a-Genome</span><ul><li><a class="toctext" href="../../build-a-genome/">Core concepts &amp; workflow</a></li><li><span class="toctext">Examples</span></li></ul></li><li><span class="toctext">Sequencing</span><ul><li><a class="toctext" href="../../sequencing/">Core concepts &amp; workflow</a></li><li><span class="toctext">Examples</span><ul><li class="current"><a class="toctext" href>Paired end reads</a><ul class="internal"><li><a class="toctext" href="#Using-the-sequence-method-1">Using the <code>sequence</code> method</a></li><li><a class="toctext" href="#Using-the-Pseudoseq-API-1">Using the Pseudoseq API</a></li></ul></li><li><a class="toctext" href="../se-example/">Long single end reads</a></li><li><a class="toctext" href="../tg-example/">Tagged paired end reads</a></li></ul></li></ul></li><li><span class="toctext">API</span><ul><li><a class="toctext" href="../../api/sequence/">Sequencing</a></li><li><a class="toctext" href="../../api/pool/">Molecule Pool</a></li><li><a class="toctext" href="../../api/reads/">Reads</a></li><li><a class="toctext" href="../../api/chromosome-blueprint/">Chromosome Blueprint</a></li></ul></li></ul></nav><article id="docs"><header><nav><ul><li>Sequencing</li><li>Examples</li><li><a href>Paired end reads</a></li></ul><a class="edit-page" href="https://github.com/bioinfologics/Pseudoseq.jl/blob/master/examples/pe-example.jl"><span class="fa"></span> Edit on GitHub</a></nav><hr/><div id="topbar"><span>Paired end reads</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="Example:-paired-end-sequencing-1" href="#Example:-paired-end-sequencing-1">Example: paired-end sequencing</a></h1><p>This is an example generated from this source file: <a href="https://github.com/bioinfologics/Pseudoseq.jl/blob/master/examples/pe-example.jl"><code>pe-example.jl</code></a> You are seeing the online documentation version. The corresponding notebook can be found here: <a href="https://nbviewer.jupyter.org/github/bioinfologics/Pseudoseq.jl/blob/gh-pages/dev/examples/pe-example.ipynb"><code>pe-example.ipynb</code></a> and the script can be found here: <a href="../pe-example.jl"><code>pe-example.jl</code></a></p><p>For the simulation we are going to:</p><ol><li>Create a pool of 5000 copies of a reference genome.</li><li>Fragment the DNA molecules in the pool, to an average length of 700bp.</li><li>Subsample the molecules in the pool to achieve approximatly 50x coverage.</li><li>Create a set of 250bp paired-end reads.</li><li>Apply errors to the paired-end reads at a rate of 0.001 (.1%).</li><li>Generate an output FASTQ file.</li></ol><div><pre><code class="language-julia">using Pseudoseq</code></pre></div><h2><a class="nav-anchor" id="Using-the-sequence-method-1" href="#Using-the-sequence-method-1">Using the <code>sequence</code> method</a></h2><p>First, let&#39;s see how we do this with the <code>sequence</code> method. The first two parameters we give to the function will be the input genome we want to sequence, and the destination FASTQ file for output reads. Here we are setting:</p><ul><li>The number of genome copies in the molecule pool to 5000.</li><li>The average fragment size to 700bp.</li><li>The sampling coverage to 50x.</li><li>The read length to 250bp.</li><li>The per base read error rate to 0.001.</li><li>The fact we want paired-ends of fragments to be read (<code>paired</code>) to true.</li></ul><div><pre><code class="language-julia">sequence(&quot;ecoli-ref.fasta&quot;, &quot;pe-reads.fastq&quot;; ng = 5000, flen = 700, cov = 50, paired = true, rdlen = 250, err = 0.001)</code></pre><pre><code class="language-none">[ Info: - ✔ Created pool of 5000 copies of a 4639675bp genome
[ Info: - ✔ Created pool of fragments with an average length of 700bp
[ Info: - ✔ Subsampled pool at 50X coverage (463967 molecules)
[ Info: - ✔ Created set of 250bp paired-end reads
[ Info: - ✔ Applied sequencing errors at a per-base rate of 0.001
[ Info: - ✔ Wrote 649702 paired end reads to FASTQ file</code></pre></div><h2><a class="nav-anchor" id="Using-the-Pseudoseq-API-1" href="#Using-the-Pseudoseq-API-1">Using the Pseudoseq API</a></h2><p>Here&#39;s how to achieve the same thing, using the Pseudoseq API. It is nessecery to use the API if you want to do something that is not anticipated by the available functionality of the <code>sequence</code> method: the cost of conveinience is fewer options.</p><p>Starting with a FASTA formatted file containing the genome we want to sequence, we create a pool with 5000 copies of the genome.</p><div><pre><code class="language-julia">pool = makepool(&quot;ecoli-ref.fasta&quot;, 5000)</code></pre><pre><code class="language-none">Pool of 5000 molecules:
 All molecules are of the same size: 4639675</code></pre></div><p>Next we use the fragment function to make a pool of shorter DNA molecules.</p><div><pre><code class="language-julia">cutpool = fragment(pool, 700)</code></pre><pre><code class="language-none">Pool of 33139992 molecules:
 Maximum molecule size: 11671
 Average molecule size: 700
 Minimum molecule size: 1</code></pre></div><p>We need to determine the number of molecules to sample, and subsample the pool:</p><div><pre><code class="language-julia">genome_size = 4639675
expected_coverage = 50
read_length = 250

N = needed_sample_size(expected_coverage, genome_size, read_length)
N = div(N, 2) # Divide by 2 as we&#39;re doing paired end sequencing.

sampledpool = subsample(cutpool, N)</code></pre><pre><code class="language-none">Pool of 463967 molecules:
 Maximum molecule size: 8785
 Average molecule size: 698
 Minimum molecule size: 1</code></pre></div><p>We now want to create a set of paired-end reads. We want our reads to be 250bp in length.</p><div><pre><code class="language-julia">pe_reads = make_reads(PairedEnd, sampledpool, 250)</code></pre></div><p>Now we have some reads, we should mark positions in the reads that are destined to be errors in the output FASTQ.</p><div><pre><code class="language-julia">pe_w_errs = mark_errors(pe_reads, 0.001)</code></pre></div><p>Now we have some paired end reads and have marked some positions as errors, we can generate FASTQ files.</p><div><pre><code class="language-julia">generate(&quot;pe-reads.fastq&quot;, pe_w_errs)#-</code></pre><pre><code class="language-none">[ Info: - ✔ Wrote 649852 paired end reads to FASTQ file</code></pre></div><p><em>This page was generated using <a href="https://github.com/fredrikekre/Literate.jl">Literate.jl</a>.</em></p><footer><hr/><a class="previous" href="../../sequencing/"><span class="direction">Previous</span><span class="title">Core concepts &amp; workflow</span></a><a class="next" href="../se-example/"><span class="direction">Next</span><span class="title">Long single end reads</span></a></footer></article></body></html>
